{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9c0fb474-f82e-46fe-9eb0-8ffb1bb6fb5a",
   "metadata": {},
   "source": [
    "## Notebook to format and re-annotate the IPDGC Meta5 ([Nalls PD GWAS 2019](https://pubmed.ncbi.nlm.nih.gov/31701892/), [without 23andMe data](https://drive.google.com/file/d/1FZ9UL99LAqyWnyNBxxlx6qOUlfAnublN/view?usp=sharing)) summary stats pulled from pdgenetics.org \n",
    "\n",
    "- summary stats as downloaded have position based IDs based on hg19 reference\n",
    "- annotate with dbSNP IDs\n",
    "- annotate with hg38 positions\n",
    "- modify effect allele and frequency info such that the minor allele is the effect allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2c27c062-503e-402b-8915-aab063ff483f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Jul 29 16:44:57 UTC 2022\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a282fb0-c95b-4b73-a73e-3070c87852fb",
   "metadata": {},
   "source": [
    "#### import libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a735730b-f9a3-4387-a933-751242a3c310",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pandas import read_csv, DataFrame\n",
    "from os import makedirs\n",
    "import concurrent.futures\n",
    "from dask.dataframe import read_csv as dask_read_csv\n",
    "from numpy import exp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a62a6366-2140-4edd-abcc-6af22d21b048",
   "metadata": {},
   "source": [
    "#### set notebook variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "db0f668c-90f8-4541-a01b-19f62b5580df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# naming\n",
    "project = 'adrd'\n",
    "cohort = 'ipsc'\n",
    "\n",
    "# directories\n",
    "wrk_dir = '/home/jupyter/ipsc_qtl'\n",
    "public_dir = f'{wrk_dir}/public'\n",
    "\n",
    "# in files\n",
    "in_sum_stats_file = f'{public_dir}/nalls_pd_gwas/nallsEtAl2019_excluding23andMe_allVariants.tab.zip'\n",
    "hrc_file = f'{public_dir}/HRC.r1-1.GRCh37.wgs.mac5.sites.tab.gz'\n",
    "\n",
    "# out files\n",
    "out_file = f'{public_dir}/nalls_pd_gwas/pdmeta_sumstats_hg38_no23andme.parquet'\n",
    "out_catalogue_file = f'{public_dir}/nalls_pd_gwas/pd_sumstats_no23andme_buildGRCh38.tsv.gz'\n",
    "\n",
    "# variables\n",
    "DEBUG = True\n",
    "autosomes = [str(x) for x in list(range(1,23))]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48f674a8-e5e7-445f-9969-754fa76e5149",
   "metadata": {},
   "source": [
    "### load input data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a77e4e5-03ab-4b3e-a47c-db4f721bc4ae",
   "metadata": {},
   "source": [
    "#### load the GWAS summary stats file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "75827059-1336-4e74-9e00-590d38f934d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gwas summary stats shape (17510617, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>freq</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>p</th>\n",
       "      <th>N_cases</th>\n",
       "      <th>N_controls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9931</td>\n",
       "      <td>0.1575</td>\n",
       "      <td>0.1783</td>\n",
       "      <td>0.3771</td>\n",
       "      <td>7161</td>\n",
       "      <td>5356</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9336</td>\n",
       "      <td>0.0605</td>\n",
       "      <td>0.0456</td>\n",
       "      <td>0.1846</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2:18069070</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9988</td>\n",
       "      <td>-0.6774</td>\n",
       "      <td>1.3519</td>\n",
       "      <td>0.6163</td>\n",
       "      <td>582</td>\n",
       "      <td>905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2081</td>\n",
       "      <td>-0.0358</td>\n",
       "      <td>0.0273</td>\n",
       "      <td>0.1887</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr12:3871714</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9972</td>\n",
       "      <td>0.1489</td>\n",
       "      <td>1.0636</td>\n",
       "      <td>0.8886</td>\n",
       "      <td>749</td>\n",
       "      <td>658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SNP A1 A2    freq       b      se       p  N_cases  N_controls\n",
       "0  chr11:88249377  T  C  0.9931  0.1575  0.1783  0.3771     7161        5356\n",
       "1   chr1:60320992  A  G  0.9336  0.0605  0.0456  0.1846    26421      442271\n",
       "2   chr2:18069070  T  C  0.9988 -0.6774  1.3519  0.6163      582         905\n",
       "3  chr8:135908647  A  G  0.2081 -0.0358  0.0273  0.1887    26421      442271\n",
       "4   chr12:3871714  A  C  0.9972  0.1489  1.0636  0.8886      749         658"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 19.4 s, sys: 1.7 s, total: 21.1 s\n",
      "Wall time: 21.1 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "gwas_df = read_csv(in_sum_stats_file, sep='\\t')\n",
    "print(f'gwas summary stats shape {gwas_df.shape}')\n",
    "\n",
    "if DEBUG:\n",
    "    display(gwas_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00e5c320-5c3c-4474-9316-2331dcbed99d",
   "metadata": {},
   "source": [
    "#### load the HRC reference files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0d9ac9b8-479b-4d26-9618-89ca1f5af5fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/IPython/core/magics/execution.py:1335: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code, glob, local_ns)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HRC reference shape (40405505, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>AC_EXCLUDING_1000G</th>\n",
       "      <th>AN_EXCLUDING_1000G</th>\n",
       "      <th>AF_EXCLUDING_1000G</th>\n",
       "      <th>AA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>13380</td>\n",
       "      <td>rs571093408</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>5</td>\n",
       "      <td>64940</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0</td>\n",
       "      <td>59950</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>16071</td>\n",
       "      <td>rs541172944</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>64940</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0</td>\n",
       "      <td>59950</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>16141</td>\n",
       "      <td>rs529651976</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9</td>\n",
       "      <td>64940</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>4</td>\n",
       "      <td>59950</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>16280</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>43</td>\n",
       "      <td>64940</td>\n",
       "      <td>0.000662</td>\n",
       "      <td>2</td>\n",
       "      <td>59950</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>49298</td>\n",
       "      <td>rs200943160</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>41571</td>\n",
       "      <td>64940</td>\n",
       "      <td>0.640145</td>\n",
       "      <td>38155</td>\n",
       "      <td>59950</td>\n",
       "      <td>0.636447</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM    POS           ID REF ALT     AC     AN        AF  \\\n",
       "0      1  13380  rs571093408   C   G      5  64940  0.000077   \n",
       "1      1  16071  rs541172944   G   A      8  64940  0.000123   \n",
       "2      1  16141  rs529651976   C   T      9  64940  0.000139   \n",
       "3      1  16280            .   T   C     43  64940  0.000662   \n",
       "4      1  49298  rs200943160   T   C  41571  64940  0.640145   \n",
       "\n",
       "   AC_EXCLUDING_1000G  AN_EXCLUDING_1000G  AF_EXCLUDING_1000G AA  \n",
       "0                   0               59950            0.000000  .  \n",
       "1                   0               59950            0.000000  G  \n",
       "2                   4               59950            0.000067  C  \n",
       "3                   2               59950            0.000033  T  \n",
       "4               38155               59950            0.636447  T  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 51.5 s, sys: 4.68 s, total: 56.2 s\n",
      "Wall time: 56.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "hrc_df = read_csv(hrc_file, sep='\\t')\n",
    "print(f'HRC reference shape {hrc_df.shape}')\n",
    "\n",
    "if DEBUG:\n",
    "    display(hrc_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42b77177-712f-46f8-94b5-62079fe18075",
   "metadata": {},
   "source": [
    "### annotate dbSNP IDs to summary stats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7da56cb8-af1f-4244-96df-17888c974703",
   "metadata": {},
   "source": [
    "#### create posID type in the HRC files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8d94f951-cb0c-4092-a8a7-6fa2204d9968",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HRC reference shape (40405505, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>AC_EXCLUDING_1000G</th>\n",
       "      <th>AN_EXCLUDING_1000G</th>\n",
       "      <th>AF_EXCLUDING_1000G</th>\n",
       "      <th>AA</th>\n",
       "      <th>posID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>13380</td>\n",
       "      <td>rs571093408</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>5</td>\n",
       "      <td>64940</td>\n",
       "      <td>0.000077</td>\n",
       "      <td>0</td>\n",
       "      <td>59950</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>.</td>\n",
       "      <td>chr1:13380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>16071</td>\n",
       "      <td>rs541172944</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>8</td>\n",
       "      <td>64940</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0</td>\n",
       "      <td>59950</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1:16071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>16141</td>\n",
       "      <td>rs529651976</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9</td>\n",
       "      <td>64940</td>\n",
       "      <td>0.000139</td>\n",
       "      <td>4</td>\n",
       "      <td>59950</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1:16141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>16280</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>43</td>\n",
       "      <td>64940</td>\n",
       "      <td>0.000662</td>\n",
       "      <td>2</td>\n",
       "      <td>59950</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1:16280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>49298</td>\n",
       "      <td>rs200943160</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>41571</td>\n",
       "      <td>64940</td>\n",
       "      <td>0.640145</td>\n",
       "      <td>38155</td>\n",
       "      <td>59950</td>\n",
       "      <td>0.636447</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1:49298</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM    POS           ID REF ALT     AC     AN        AF  \\\n",
       "0      1  13380  rs571093408   C   G      5  64940  0.000077   \n",
       "1      1  16071  rs541172944   G   A      8  64940  0.000123   \n",
       "2      1  16141  rs529651976   C   T      9  64940  0.000139   \n",
       "3      1  16280            .   T   C     43  64940  0.000662   \n",
       "4      1  49298  rs200943160   T   C  41571  64940  0.640145   \n",
       "\n",
       "   AC_EXCLUDING_1000G  AN_EXCLUDING_1000G  AF_EXCLUDING_1000G AA       posID  \n",
       "0                   0               59950            0.000000  .  chr1:13380  \n",
       "1                   0               59950            0.000000  G  chr1:16071  \n",
       "2                   4               59950            0.000067  C  chr1:16141  \n",
       "3                   2               59950            0.000033  T  chr1:16280  \n",
       "4               38155               59950            0.636447  T  chr1:49298  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hrc_df['posID'] = 'chr' + hrc_df['#CHROM'].astype('str') + ':' + hrc_df.POS.astype('str')\n",
    "print(f'HRC reference shape {hrc_df.shape}')\n",
    "\n",
    "if DEBUG:\n",
    "    display(hrc_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e673c07d-25aa-425c-87af-e62bcefe8e07",
   "metadata": {},
   "source": [
    "#### how many of the summary stats variants are not present in HRC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b41e460c-16d3-4732-9c16-bffde966dff0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count or posID intersect between GWAS summary stats and HRC ref is 17510617\n",
      "count of posID in GWAS summary stats but not in HRC ref is 0\n"
     ]
    }
   ],
   "source": [
    "cnt = len(set(gwas_df.SNP) & set(hrc_df.posID))\n",
    "print(f'count or posID intersect between GWAS summary stats and HRC ref is {cnt}')\n",
    "missing_cnt = len(set(gwas_df.SNP) - set(hrc_df.posID))\n",
    "print(f'count of posID in GWAS summary stats but not in HRC ref is {missing_cnt}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0a610c7-00a8-4e68-b38d-ca6282c28473",
   "metadata": {},
   "source": [
    "#### merge subset of HRC columns to summary stats keyed on podID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a2e0e84f-60fa-4bc0-a4cb-231d386424f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gwas summary stats shape (17579908, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>freq</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>p</th>\n",
       "      <th>N_cases</th>\n",
       "      <th>N_controls</th>\n",
       "      <th>posID</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9931</td>\n",
       "      <td>0.1575</td>\n",
       "      <td>0.1783</td>\n",
       "      <td>0.3771</td>\n",
       "      <td>7161</td>\n",
       "      <td>5356</td>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>rs11020170</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9336</td>\n",
       "      <td>0.0605</td>\n",
       "      <td>0.0456</td>\n",
       "      <td>0.1846</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2:18069070</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9988</td>\n",
       "      <td>-0.6774</td>\n",
       "      <td>1.3519</td>\n",
       "      <td>0.6163</td>\n",
       "      <td>582</td>\n",
       "      <td>905</td>\n",
       "      <td>chr2:18069070</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2081</td>\n",
       "      <td>-0.0358</td>\n",
       "      <td>0.0273</td>\n",
       "      <td>0.1887</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr12:3871714</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9972</td>\n",
       "      <td>0.1489</td>\n",
       "      <td>1.0636</td>\n",
       "      <td>0.8886</td>\n",
       "      <td>749</td>\n",
       "      <td>658</td>\n",
       "      <td>chr12:3871714</td>\n",
       "      <td>rs192908256</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SNP A1 A2    freq       b      se       p  N_cases  N_controls  \\\n",
       "0  chr11:88249377  T  C  0.9931  0.1575  0.1783  0.3771     7161        5356   \n",
       "1   chr1:60320992  A  G  0.9336  0.0605  0.0456  0.1846    26421      442271   \n",
       "2   chr2:18069070  T  C  0.9988 -0.6774  1.3519  0.6163      582         905   \n",
       "3  chr8:135908647  A  G  0.2081 -0.0358  0.0273  0.1887    26421      442271   \n",
       "4   chr12:3871714  A  C  0.9972  0.1489  1.0636  0.8886      749         658   \n",
       "\n",
       "            posID           ID REF ALT  \n",
       "0  chr11:88249377   rs11020170   T   C  \n",
       "1   chr1:60320992  rs116406626   A   G  \n",
       "2   chr2:18069070            .   T   C  \n",
       "3  chr8:135908647   rs11992603   G   A  \n",
       "4   chr12:3871714  rs192908256   A   C  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hrc_cols_to_keep = ['posID', 'ID', 'REF', 'ALT']\n",
    "gwas_df = gwas_df.merge(hrc_df[hrc_cols_to_keep], how='inner', left_on='SNP', right_on='posID')\n",
    "print(f'gwas summary stats shape {gwas_df.shape}')\n",
    "\n",
    "if DEBUG:\n",
    "    display(gwas_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d2513809-a822-4224-84a2-2b68ba262de8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.072 fraction of variants without dbSNP IDs\n"
     ]
    }
   ],
   "source": [
    "cnt_wo_dbsnp = gwas_df.loc[gwas_df.ID == '.'].shape[0]\n",
    "frac_wo_dbsnp = cnt_wo_dbsnp/gwas_df.shape[0]\n",
    "print(f'{frac_wo_dbsnp:.3f} fraction of variants without dbSNP IDs')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd1242b3-8d83-4900-8339-73a7575b985e",
   "metadata": {},
   "source": [
    "### annotate summary stats with hg38 variant positions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4874c9f-1007-4db0-8ef0-8675ebfd0771",
   "metadata": {},
   "source": [
    "#### since annotating the hg38 positions based on dbSNP IDs drop the ones that don't have a dbSNP IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a7efbcae-56ed-44fa-9734-6900f4fb0283",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gwas summary stats shape (16311706, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>freq</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>p</th>\n",
       "      <th>N_cases</th>\n",
       "      <th>N_controls</th>\n",
       "      <th>posID</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9931</td>\n",
       "      <td>0.1575</td>\n",
       "      <td>0.1783</td>\n",
       "      <td>0.3771</td>\n",
       "      <td>7161</td>\n",
       "      <td>5356</td>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>rs11020170</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9336</td>\n",
       "      <td>0.0605</td>\n",
       "      <td>0.0456</td>\n",
       "      <td>0.1846</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2081</td>\n",
       "      <td>-0.0358</td>\n",
       "      <td>0.0273</td>\n",
       "      <td>0.1887</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr12:3871714</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9972</td>\n",
       "      <td>0.1489</td>\n",
       "      <td>1.0636</td>\n",
       "      <td>0.8886</td>\n",
       "      <td>749</td>\n",
       "      <td>658</td>\n",
       "      <td>chr12:3871714</td>\n",
       "      <td>rs192908256</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9976</td>\n",
       "      <td>-0.1213</td>\n",
       "      <td>0.3874</td>\n",
       "      <td>0.7543</td>\n",
       "      <td>6248</td>\n",
       "      <td>4391</td>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>rs189372368</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SNP A1 A2    freq       b      se       p  N_cases  N_controls  \\\n",
       "0  chr11:88249377  T  C  0.9931  0.1575  0.1783  0.3771     7161        5356   \n",
       "1   chr1:60320992  A  G  0.9336  0.0605  0.0456  0.1846    26421      442271   \n",
       "3  chr8:135908647  A  G  0.2081 -0.0358  0.0273  0.1887    26421      442271   \n",
       "4   chr12:3871714  A  C  0.9972  0.1489  1.0636  0.8886      749         658   \n",
       "5  chr16:77148858  A  G  0.9976 -0.1213  0.3874  0.7543     6248        4391   \n",
       "\n",
       "            posID           ID REF ALT  \n",
       "0  chr11:88249377   rs11020170   T   C  \n",
       "1   chr1:60320992  rs116406626   A   G  \n",
       "3  chr8:135908647   rs11992603   G   A  \n",
       "4   chr12:3871714  rs192908256   A   C  \n",
       "5  chr16:77148858  rs189372368   A   G  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gwas_df = gwas_df.loc[gwas_df.ID != '.']\n",
    "print(f'gwas summary stats shape {gwas_df.shape}')\n",
    "\n",
    "if DEBUG:\n",
    "    display(gwas_df.head())    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "712ebde4-1b02-4b23-9db8-47d2f127be38",
   "metadata": {},
   "source": [
    "#### pull down the dbSNP annotation files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "12f77809-e60b-4e18-a3ef-9392b387a7e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_bash_cmd(cmd: str):\n",
    "    !{cmd}\n",
    "\n",
    "#funciton to pull down the UCSC dbSNP files\n",
    "def pull_ucsc_dbsnp_files(version_build_name: str, chrs_list: list, \n",
    "                          out_dir: str, threaded: bool=False, verbose: bool=False):\n",
    "    snp_bed_url = f'ftp://ftp.ncbi.nih.gov/snp/organisms/{version_build_name}/BED/'\n",
    "    with concurrent.futures.ProcessPoolExecutor() as ppe:\n",
    "        for chrom in chrs_list:\n",
    "            this_chr_snp_bed = f'bed_chr_{chrom}.bed.gz'\n",
    "            snp_bed_file = snp_bed_url + this_chr_snp_bed\n",
    "            this_cmd = f'curl --silent -L {snp_bed_file} --output {out_dir}/{this_chr_snp_bed}'\n",
    "            if verbose:\n",
    "                print(this_cmd)\n",
    "            if threaded:\n",
    "                ppe.submit(run_bash_cmd, this_cmd)\n",
    "            else:\n",
    "                run_bash_cmd(this_cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b0606750-a2a3-45a4-bd41-37fb879cc31e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 64.2 ms, sys: 3.53 s, total: 3.59 s\n",
      "Wall time: 1min 9s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# pull down hg38 dbSNP b151\n",
    "out_dir = f'{public_dir}/dbsnp_hg38'\n",
    "makedirs(out_dir, exist_ok=True)\n",
    "\n",
    "pull_ucsc_dbsnp_files('human_9606_b151_GRCh38p7', autosomes, out_dir, \n",
    "                      threaded=True, verbose=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffdc5fde-7f48-4d3b-bfec-8e80f4744750",
   "metadata": {},
   "source": [
    "#### load the dbSNP bed files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d9778309-f7d6-4a55-a481-a3d6529a6ef9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5 µs, sys: 0 ns, total: 5 µs\n",
      "Wall time: 7.63 µs\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# def load_dbsnp_bed(file_path: str) -> DataFrame:\n",
    "#     this_df = read_csv(file_path, skiprows=1, header=None, sep='\\t')\n",
    "#     this_df.drop(columns=[1, 4, 5], inplace=True)\n",
    "#     this_df.columns = ['chrom', 'position', 'id']\n",
    "#     return this_df\n",
    "\n",
    "# fs_list = []\n",
    "# lm_results = []\n",
    "# with concurrent.futures.ThreadPoolExecutor() as tpe:\n",
    "#     for chrom in autosomes:\n",
    "#         this_file = f'{public_dir}/dbsnp_hg38/bed_chr_{chrom}.bed.gz'\n",
    "#         fs_list.append(tpe.submit(load_dbsnp_bed, this_file))\n",
    "# for future in concurrent.futures.as_completed(fs_list):\n",
    "#     load_results.append(future.result())\n",
    "\n",
    "# # flatten the list\n",
    "# dbsnp_df = concat([item for item in load_results])\n",
    "# print(dbsnp_df.shape)\n",
    "# if DEBUG:\n",
    "#     display(dbsnp_df.sample(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bb0ae025-ee14-4c7c-8c8e-47a0acb7413f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.7/site-packages/dask/dataframe/io/csv.py:533: UserWarning: Warning gzip compression does not support breaking apart files\n",
      "Please ensure that each individual file can fit in memory and\n",
      "use the keyword ``blocksize=None to remove this message``\n",
      "Setting ``blocksize=None``\n",
      "  \"Setting ``blocksize=None``\" % compression\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(633520791, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17785105</th>\n",
       "      <td>chr3</td>\n",
       "      <td>139990223</td>\n",
       "      <td>rs989100640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5232190</th>\n",
       "      <td>chr5</td>\n",
       "      <td>130555511</td>\n",
       "      <td>rs551330192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3519084</th>\n",
       "      <td>chr19</td>\n",
       "      <td>31157075</td>\n",
       "      <td>rs781603457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22299731</th>\n",
       "      <td>chr12</td>\n",
       "      <td>453909</td>\n",
       "      <td>rs1315965319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15871892</th>\n",
       "      <td>chr14</td>\n",
       "      <td>66200769</td>\n",
       "      <td>rs1342810072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>825243</th>\n",
       "      <td>chr10</td>\n",
       "      <td>15901476</td>\n",
       "      <td>rs77388106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33749671</th>\n",
       "      <td>chr1</td>\n",
       "      <td>212281844</td>\n",
       "      <td>rs1265157675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12210849</th>\n",
       "      <td>chr6</td>\n",
       "      <td>98548145</td>\n",
       "      <td>rs939358627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367880</th>\n",
       "      <td>chr7</td>\n",
       "      <td>64456505</td>\n",
       "      <td>rs527562885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44367662</th>\n",
       "      <td>chr3</td>\n",
       "      <td>69373073</td>\n",
       "      <td>rs1476055991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          chrom   position            id\n",
       "17785105   chr3  139990223   rs989100640\n",
       "5232190    chr5  130555511   rs551330192\n",
       "3519084   chr19   31157075   rs781603457\n",
       "22299731  chr12     453909  rs1315965319\n",
       "15871892  chr14   66200769  rs1342810072\n",
       "825243    chr10   15901476    rs77388106\n",
       "33749671   chr1  212281844  rs1265157675\n",
       "12210849   chr6   98548145   rs939358627\n",
       "3367880    chr7   64456505   rs527562885\n",
       "44367662   chr3   69373073  rs1476055991"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 14min 3s, sys: 4min 1s, total: 18min 4s\n",
      "Wall time: 9min 29s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "bed_files = f'{public_dir}/dbsnp_hg38/bed_chr_*.bed.gz'\n",
    "dbsnp_df = dask_read_csv(bed_files, skiprows=1, header=None, \n",
    "                         sep='\\t', usecols=[0, 2, 3]).compute()\n",
    "# dbsnp_df.drop(columns=[1, 4, 5], inplace=True)\n",
    "dbsnp_df.columns = ['chrom', 'position', 'id']\n",
    "print(dbsnp_df.shape)\n",
    "if DEBUG:\n",
    "    display(dbsnp_df.sample(10))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8a7ee88-2d9c-402f-a190-cba654121cb8",
   "metadata": {},
   "source": [
    "#### only need the variants that are in the summary stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "df58908d-1f9d-404d-a4bb-07ed8998349d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(16263779, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3222549</th>\n",
       "      <td>chr14</td>\n",
       "      <td>60098879</td>\n",
       "      <td>rs569515776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2600502</th>\n",
       "      <td>chr2</td>\n",
       "      <td>164269495</td>\n",
       "      <td>rs147574942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>794877</th>\n",
       "      <td>chr6</td>\n",
       "      <td>5039553</td>\n",
       "      <td>rs62385159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408496</th>\n",
       "      <td>chr8</td>\n",
       "      <td>29591134</td>\n",
       "      <td>rs17060380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>886211</th>\n",
       "      <td>chr3</td>\n",
       "      <td>98168032</td>\n",
       "      <td>rs72931982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5977525</th>\n",
       "      <td>chr5</td>\n",
       "      <td>118008414</td>\n",
       "      <td>rs562831561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1190309</th>\n",
       "      <td>chr4</td>\n",
       "      <td>101284781</td>\n",
       "      <td>rs78650301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439314</th>\n",
       "      <td>chr1</td>\n",
       "      <td>39041489</td>\n",
       "      <td>rs10888650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284481</th>\n",
       "      <td>chr5</td>\n",
       "      <td>58983004</td>\n",
       "      <td>rs7713345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310263</th>\n",
       "      <td>chr4</td>\n",
       "      <td>56483116</td>\n",
       "      <td>rs7700034</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         chrom   position           id\n",
       "3222549  chr14   60098879  rs569515776\n",
       "2600502   chr2  164269495  rs147574942\n",
       "794877    chr6    5039553   rs62385159\n",
       "408496    chr8   29591134   rs17060380\n",
       "886211    chr3   98168032   rs72931982\n",
       "5977525   chr5  118008414  rs562831561\n",
       "1190309   chr4  101284781   rs78650301\n",
       "439314    chr1   39041489   rs10888650\n",
       "284481    chr5   58983004    rs7713345\n",
       "310263    chr4   56483116    rs7700034"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dbsnp_df = dbsnp_df.loc[dbsnp_df.id.isin(gwas_df.ID)]\n",
    "print(dbsnp_df.shape)\n",
    "if DEBUG:\n",
    "    display(dbsnp_df.sample(10))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a59531b-4a6c-4b19-b35e-1f9ac8f34d40",
   "metadata": {},
   "source": [
    "#### merget the dbSNP 151 hg38 annotations onto the gwas summary statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "966b8442-e286-4bc6-9eb2-efd780183ea0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gwas summary stats shape (16311706, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>freq</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>p</th>\n",
       "      <th>N_cases</th>\n",
       "      <th>N_controls</th>\n",
       "      <th>posID</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9931</td>\n",
       "      <td>0.1575</td>\n",
       "      <td>0.1783</td>\n",
       "      <td>0.3771</td>\n",
       "      <td>7161</td>\n",
       "      <td>5356</td>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>rs11020170</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr11</td>\n",
       "      <td>88516209.0</td>\n",
       "      <td>rs11020170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9336</td>\n",
       "      <td>0.0605</td>\n",
       "      <td>0.0456</td>\n",
       "      <td>0.1846</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1</td>\n",
       "      <td>59855320.0</td>\n",
       "      <td>rs116406626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2081</td>\n",
       "      <td>-0.0358</td>\n",
       "      <td>0.0273</td>\n",
       "      <td>0.1887</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr8</td>\n",
       "      <td>134896404.0</td>\n",
       "      <td>rs11992603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12:3871714</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9972</td>\n",
       "      <td>0.1489</td>\n",
       "      <td>1.0636</td>\n",
       "      <td>0.8886</td>\n",
       "      <td>749</td>\n",
       "      <td>658</td>\n",
       "      <td>chr12:3871714</td>\n",
       "      <td>rs192908256</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr12</td>\n",
       "      <td>3762548.0</td>\n",
       "      <td>rs192908256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9976</td>\n",
       "      <td>-0.1213</td>\n",
       "      <td>0.3874</td>\n",
       "      <td>0.7543</td>\n",
       "      <td>6248</td>\n",
       "      <td>4391</td>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>rs189372368</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr16</td>\n",
       "      <td>77114961.0</td>\n",
       "      <td>rs189372368</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SNP A1 A2    freq       b      se       p  N_cases  N_controls  \\\n",
       "0  chr11:88249377  T  C  0.9931  0.1575  0.1783  0.3771     7161        5356   \n",
       "1   chr1:60320992  A  G  0.9336  0.0605  0.0456  0.1846    26421      442271   \n",
       "2  chr8:135908647  A  G  0.2081 -0.0358  0.0273  0.1887    26421      442271   \n",
       "3   chr12:3871714  A  C  0.9972  0.1489  1.0636  0.8886      749         658   \n",
       "4  chr16:77148858  A  G  0.9976 -0.1213  0.3874  0.7543     6248        4391   \n",
       "\n",
       "            posID           ID REF ALT  chrom     position           id  \n",
       "0  chr11:88249377   rs11020170   T   C  chr11   88516209.0   rs11020170  \n",
       "1   chr1:60320992  rs116406626   A   G   chr1   59855320.0  rs116406626  \n",
       "2  chr8:135908647   rs11992603   G   A   chr8  134896404.0   rs11992603  \n",
       "3   chr12:3871714  rs192908256   A   C  chr12    3762548.0  rs192908256  \n",
       "4  chr16:77148858  rs189372368   A   G  chr16   77114961.0  rs189372368  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gwas_df = gwas_df.merge(dbsnp_df, how='left', left_on='ID', right_on='id')\n",
    "print(f'gwas summary stats shape {gwas_df.shape}')\n",
    "\n",
    "if DEBUG:\n",
    "    display(gwas_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3aedd64-a315-49cf-bc9b-6955407b0f51",
   "metadata": {},
   "source": [
    "#### how many of the the HRC dbSNP IDs aren't in or didn't merge for some other resion in dbSNP151\n",
    "note could've just used 'inner' merge to not include these but doing in two steps so I can see the count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0895baaf-834a-48a9-ac2e-a49de57513a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "number of variants that did not merge from dbSNP 151 18118\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>freq</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>p</th>\n",
       "      <th>N_cases</th>\n",
       "      <th>N_controls</th>\n",
       "      <th>posID</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>549</th>\n",
       "      <td>chr6:29968965</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0640</td>\n",
       "      <td>0.0493</td>\n",
       "      <td>0.0360</td>\n",
       "      <td>0.17170</td>\n",
       "      <td>32505</td>\n",
       "      <td>448088</td>\n",
       "      <td>chr6:29968965</td>\n",
       "      <td>rs558185119</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1546</th>\n",
       "      <td>chr2:129706224</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9989</td>\n",
       "      <td>-0.0673</td>\n",
       "      <td>0.6657</td>\n",
       "      <td>0.91950</td>\n",
       "      <td>2692</td>\n",
       "      <td>2238</td>\n",
       "      <td>chr2:129706224</td>\n",
       "      <td>rs111489065</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2273</th>\n",
       "      <td>chr10:107779570</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2610</td>\n",
       "      <td>-0.0558</td>\n",
       "      <td>0.0307</td>\n",
       "      <td>0.06927</td>\n",
       "      <td>7803</td>\n",
       "      <td>5852</td>\n",
       "      <td>chr10:107779570</td>\n",
       "      <td>rs112715238</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3790</th>\n",
       "      <td>chr14:74592024</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.8584</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0.0251</td>\n",
       "      <td>0.91300</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>chr14:74592024</td>\n",
       "      <td>rs138656424</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3984</th>\n",
       "      <td>chr21:18617763</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.1024</td>\n",
       "      <td>0.0046</td>\n",
       "      <td>0.0422</td>\n",
       "      <td>0.91220</td>\n",
       "      <td>25252</td>\n",
       "      <td>441303</td>\n",
       "      <td>chr21:18617763</td>\n",
       "      <td>rs146412710</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  SNP A1 A2    freq       b      se        p  N_cases  \\\n",
       "549     chr6:29968965  A  G  0.0640  0.0493  0.0360  0.17170    32505   \n",
       "1546   chr2:129706224  T  C  0.9989 -0.0673  0.6657  0.91950     2692   \n",
       "2273  chr10:107779570  A  G  0.2610 -0.0558  0.0307  0.06927     7803   \n",
       "3790   chr14:74592024  A  G  0.8584  0.0027  0.0251  0.91300    33674   \n",
       "3984   chr21:18617763  A  C  0.1024  0.0046  0.0422  0.91220    25252   \n",
       "\n",
       "      N_controls            posID           ID REF ALT chrom  position   id  \n",
       "549       448088    chr6:29968965  rs558185119   G   A   NaN       NaN  NaN  \n",
       "1546        2238   chr2:129706224  rs111489065   C   T   NaN       NaN  NaN  \n",
       "2273        5852  chr10:107779570  rs112715238   G   A   NaN       NaN  NaN  \n",
       "3790      449056   chr14:74592024  rs138656424   A   G   NaN       NaN  NaN  \n",
       "3984      441303   chr21:18617763  rs146412710   C   A   NaN       NaN  NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "missing_df = gwas_df.loc[gwas_df.ID != gwas_df.id]\n",
    "print(f'number of variants that did not merge from dbSNP 151 {missing_df.shape[0]}')\n",
    "if DEBUG:\n",
    "    display(missing_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccedd529-f8b6-4866-840d-f54299f9c07d",
   "metadata": {},
   "source": [
    "#### go ahead and remove the ones that were missing from dbSNP 151"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "96c483e7-4254-4079-8743-2db3d74dd0bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gwas summary stats shape (16293588, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>freq</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>p</th>\n",
       "      <th>N_cases</th>\n",
       "      <th>N_controls</th>\n",
       "      <th>posID</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9931</td>\n",
       "      <td>0.1575</td>\n",
       "      <td>0.1783</td>\n",
       "      <td>0.3771</td>\n",
       "      <td>7161</td>\n",
       "      <td>5356</td>\n",
       "      <td>chr11:88249377</td>\n",
       "      <td>rs11020170</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr11</td>\n",
       "      <td>88516209.0</td>\n",
       "      <td>rs11020170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9336</td>\n",
       "      <td>0.0605</td>\n",
       "      <td>0.0456</td>\n",
       "      <td>0.1846</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>chr1:60320992</td>\n",
       "      <td>rs116406626</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr1</td>\n",
       "      <td>59855320.0</td>\n",
       "      <td>rs116406626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2081</td>\n",
       "      <td>-0.0358</td>\n",
       "      <td>0.0273</td>\n",
       "      <td>0.1887</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>chr8:135908647</td>\n",
       "      <td>rs11992603</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr8</td>\n",
       "      <td>134896404.0</td>\n",
       "      <td>rs11992603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12:3871714</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9972</td>\n",
       "      <td>0.1489</td>\n",
       "      <td>1.0636</td>\n",
       "      <td>0.8886</td>\n",
       "      <td>749</td>\n",
       "      <td>658</td>\n",
       "      <td>chr12:3871714</td>\n",
       "      <td>rs192908256</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr12</td>\n",
       "      <td>3762548.0</td>\n",
       "      <td>rs192908256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9976</td>\n",
       "      <td>-0.1213</td>\n",
       "      <td>0.3874</td>\n",
       "      <td>0.7543</td>\n",
       "      <td>6248</td>\n",
       "      <td>4391</td>\n",
       "      <td>chr16:77148858</td>\n",
       "      <td>rs189372368</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr16</td>\n",
       "      <td>77114961.0</td>\n",
       "      <td>rs189372368</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              SNP A1 A2    freq       b      se       p  N_cases  N_controls  \\\n",
       "0  chr11:88249377  T  C  0.9931  0.1575  0.1783  0.3771     7161        5356   \n",
       "1   chr1:60320992  A  G  0.9336  0.0605  0.0456  0.1846    26421      442271   \n",
       "2  chr8:135908647  A  G  0.2081 -0.0358  0.0273  0.1887    26421      442271   \n",
       "3   chr12:3871714  A  C  0.9972  0.1489  1.0636  0.8886      749         658   \n",
       "4  chr16:77148858  A  G  0.9976 -0.1213  0.3874  0.7543     6248        4391   \n",
       "\n",
       "            posID           ID REF ALT  chrom     position           id  \n",
       "0  chr11:88249377   rs11020170   T   C  chr11   88516209.0   rs11020170  \n",
       "1   chr1:60320992  rs116406626   A   G   chr1   59855320.0  rs116406626  \n",
       "2  chr8:135908647   rs11992603   G   A   chr8  134896404.0   rs11992603  \n",
       "3   chr12:3871714  rs192908256   A   C  chr12    3762548.0  rs192908256  \n",
       "4  chr16:77148858  rs189372368   A   G  chr16   77114961.0  rs189372368  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gwas_df = gwas_df.loc[~gwas_df.id.isna()]\n",
    "print(f'gwas summary stats shape {gwas_df.shape}')\n",
    "\n",
    "if DEBUG:\n",
    "    display(gwas_df.head()) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "32b9c101-2dde-42aa-83f1-f0acb0f42482",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>freq</th>\n",
       "      <th>b</th>\n",
       "      <th>se</th>\n",
       "      <th>p</th>\n",
       "      <th>N_cases</th>\n",
       "      <th>N_controls</th>\n",
       "      <th>posID</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>chrom</th>\n",
       "      <th>position</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2764250</th>\n",
       "      <td>chr1:205723572</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.5748</td>\n",
       "      <td>0.0999</td>\n",
       "      <td>0.0171</td>\n",
       "      <td>4.941000e-09</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>chr1:205723572</td>\n",
       "      <td>rs823118</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1</td>\n",
       "      <td>205754444.0</td>\n",
       "      <td>rs823118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3539407</th>\n",
       "      <td>chr1:205737739</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.1426</td>\n",
       "      <td>0.1351</td>\n",
       "      <td>0.0241</td>\n",
       "      <td>2.122000e-08</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>chr1:205737739</td>\n",
       "      <td>rs11557080</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>205768611.0</td>\n",
       "      <td>rs11557080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3893676</th>\n",
       "      <td>chr1:154898185</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0112</td>\n",
       "      <td>0.3113</td>\n",
       "      <td>0.0844</td>\n",
       "      <td>2.246000e-04</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>chr1:154898185</td>\n",
       "      <td>rs114138760</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1</td>\n",
       "      <td>154925709.0</td>\n",
       "      <td>rs114138760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6114374</th>\n",
       "      <td>chr1:155205634</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9930</td>\n",
       "      <td>-0.4907</td>\n",
       "      <td>0.1426</td>\n",
       "      <td>5.764000e-04</td>\n",
       "      <td>12989</td>\n",
       "      <td>10894</td>\n",
       "      <td>chr1:155205634</td>\n",
       "      <td>rs76763715</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843.0</td>\n",
       "      <td>rs76763715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6568248</th>\n",
       "      <td>chr2:18147848</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9105</td>\n",
       "      <td>0.1551</td>\n",
       "      <td>0.0403</td>\n",
       "      <td>1.190000e-04</td>\n",
       "      <td>26948</td>\n",
       "      <td>442743</td>\n",
       "      <td>chr2:18147848</td>\n",
       "      <td>rs76116224</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr2</td>\n",
       "      <td>17966582.0</td>\n",
       "      <td>rs76116224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7573730</th>\n",
       "      <td>chr1:171719769</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.1955</td>\n",
       "      <td>-0.0781</td>\n",
       "      <td>0.0222</td>\n",
       "      <td>4.235000e-04</td>\n",
       "      <td>32505</td>\n",
       "      <td>448088</td>\n",
       "      <td>chr1:171719769</td>\n",
       "      <td>rs11578699</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1</td>\n",
       "      <td>171750629.0</td>\n",
       "      <td>rs11578699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9698034</th>\n",
       "      <td>chr1:161469054</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.5014</td>\n",
       "      <td>0.0722</td>\n",
       "      <td>0.0171</td>\n",
       "      <td>2.418000e-05</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>chr1:161469054</td>\n",
       "      <td>rs6658353</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1</td>\n",
       "      <td>161499264.0</td>\n",
       "      <td>rs6658353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10185847</th>\n",
       "      <td>chr1:226916078</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.7158</td>\n",
       "      <td>0.0729</td>\n",
       "      <td>0.0186</td>\n",
       "      <td>8.667000e-05</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>chr1:226916078</td>\n",
       "      <td>rs4653767</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr1</td>\n",
       "      <td>226728377.0</td>\n",
       "      <td>rs4653767</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13526195</th>\n",
       "      <td>chr1:155135036</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0191</td>\n",
       "      <td>0.7508</td>\n",
       "      <td>0.0659</td>\n",
       "      <td>5.022000e-30</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>chr1:155135036</td>\n",
       "      <td>rs35749011</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155162560.0</td>\n",
       "      <td>rs35749011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15191523</th>\n",
       "      <td>chr1:232664611</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.1434</td>\n",
       "      <td>0.0998</td>\n",
       "      <td>0.0241</td>\n",
       "      <td>3.532000e-05</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>chr1:232664611</td>\n",
       "      <td>rs10797576</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1</td>\n",
       "      <td>232528865.0</td>\n",
       "      <td>rs10797576</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     SNP A1 A2    freq       b      se             p  N_cases  \\\n",
       "2764250   chr1:205723572  T  C  0.5748  0.0999  0.0171  4.941000e-09    33674   \n",
       "3539407   chr1:205737739  A  G  0.1426  0.1351  0.0241  2.122000e-08    33674   \n",
       "3893676   chr1:154898185  C  G  0.0112  0.3113  0.0844  2.246000e-04    33674   \n",
       "6114374   chr1:155205634  T  C  0.9930 -0.4907  0.1426  5.764000e-04    12989   \n",
       "6568248    chr2:18147848  A  T  0.9105  0.1551  0.0403  1.190000e-04    26948   \n",
       "7573730   chr1:171719769  T  C  0.1955 -0.0781  0.0222  4.235000e-04    32505   \n",
       "9698034   chr1:161469054  C  G  0.5014  0.0722  0.0171  2.418000e-05    33674   \n",
       "10185847  chr1:226916078  T  C  0.7158  0.0729  0.0186  8.667000e-05    33674   \n",
       "13526195  chr1:155135036  A  G  0.0191  0.7508  0.0659  5.022000e-30    33674   \n",
       "15191523  chr1:232664611  T  C  0.1434  0.0998  0.0241  3.532000e-05    33674   \n",
       "\n",
       "          N_controls           posID           ID REF ALT chrom     position  \\\n",
       "2764250       449056  chr1:205723572     rs823118   C   T  chr1  205754444.0   \n",
       "3539407       449056  chr1:205737739   rs11557080   G   A  chr1  205768611.0   \n",
       "3893676       449056  chr1:154898185  rs114138760   G   C  chr1  154925709.0   \n",
       "6114374        10894  chr1:155205634   rs76763715   T   C  chr1  155235843.0   \n",
       "6568248       442743   chr2:18147848   rs76116224   A   T  chr2   17966582.0   \n",
       "7573730       448088  chr1:171719769   rs11578699   C   T  chr1  171750629.0   \n",
       "9698034       449056  chr1:161469054    rs6658353   G   C  chr1  161499264.0   \n",
       "10185847      449056  chr1:226916078    rs4653767   T   C  chr1  226728377.0   \n",
       "13526195      449056  chr1:155135036   rs35749011   G   A  chr1  155162560.0   \n",
       "15191523      449056  chr1:232664611   rs10797576   C   T  chr1  232528865.0   \n",
       "\n",
       "                   id  \n",
       "2764250      rs823118  \n",
       "3539407    rs11557080  \n",
       "3893676   rs114138760  \n",
       "6114374    rs76763715  \n",
       "6568248    rs76116224  \n",
       "7573730    rs11578699  \n",
       "9698034     rs6658353  \n",
       "10185847    rs4653767  \n",
       "13526195   rs35749011  \n",
       "15191523   rs10797576  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "test_snps = ['rs114138760', 'rs35749011', 'rs76763715', 'rs6658353',\n",
    "             'rs11578699', 'rs823118', 'rs11557080', 'rs4653767',\n",
    "             'rs10797576', 'rs76116224']\n",
    "display(gwas_df.loc[gwas_df.ID.isin(test_snps)])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d5e4c6c-b999-4cac-aa68-51857c8dbc1f",
   "metadata": {},
   "source": [
    "#### cleanup, re-order, and subset gwas summary statistis to keep\n",
    "format columns to be very similar to public summary stats available via NHGRI-EBI GWAS Catalog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f2f8d25e-c902-487b-ac71-e7132c57ea01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gwas summary stats shape (16293588, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>p_value</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>base_pair_location</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>effect_allele_frequency</th>\n",
       "      <th>beta</th>\n",
       "      <th>standard_error</th>\n",
       "      <th>n_cases</th>\n",
       "      <th>n_controls</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs11020170</td>\n",
       "      <td>0.3771</td>\n",
       "      <td>11</td>\n",
       "      <td>88516209</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9931</td>\n",
       "      <td>0.1575</td>\n",
       "      <td>0.1783</td>\n",
       "      <td>7161</td>\n",
       "      <td>5356</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs116406626</td>\n",
       "      <td>0.1846</td>\n",
       "      <td>1</td>\n",
       "      <td>59855320</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9336</td>\n",
       "      <td>0.0605</td>\n",
       "      <td>0.0456</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs11992603</td>\n",
       "      <td>0.1887</td>\n",
       "      <td>8</td>\n",
       "      <td>134896404</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2081</td>\n",
       "      <td>-0.0358</td>\n",
       "      <td>0.0273</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs192908256</td>\n",
       "      <td>0.8886</td>\n",
       "      <td>12</td>\n",
       "      <td>3762548</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9972</td>\n",
       "      <td>0.1489</td>\n",
       "      <td>1.0636</td>\n",
       "      <td>749</td>\n",
       "      <td>658</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs189372368</td>\n",
       "      <td>0.7543</td>\n",
       "      <td>16</td>\n",
       "      <td>77114961</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9976</td>\n",
       "      <td>-0.1213</td>\n",
       "      <td>0.3874</td>\n",
       "      <td>6248</td>\n",
       "      <td>4391</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id  p_value chromosome  base_pair_location effect_allele  \\\n",
       "0   rs11020170   0.3771         11            88516209             T   \n",
       "1  rs116406626   0.1846          1            59855320             A   \n",
       "2   rs11992603   0.1887          8           134896404             A   \n",
       "3  rs192908256   0.8886         12             3762548             A   \n",
       "4  rs189372368   0.7543         16            77114961             A   \n",
       "\n",
       "  other_allele  effect_allele_frequency    beta  standard_error  n_cases  \\\n",
       "0            C                   0.9931  0.1575          0.1783     7161   \n",
       "1            G                   0.9336  0.0605          0.0456    26421   \n",
       "2            G                   0.2081 -0.0358          0.0273    26421   \n",
       "3            C                   0.9972  0.1489          1.0636      749   \n",
       "4            G                   0.9976 -0.1213          0.3874     6248   \n",
       "\n",
       "   n_controls ref_allele alt_allele  \n",
       "0        5356          T          C  \n",
       "1      442271          A          G  \n",
       "2      442271          G          A  \n",
       "3         658          A          C  \n",
       "4        4391          A          G  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gwas_df.rename(columns={'ID': 'variant_id', 'p': 'p_value', 'chrom': 'chromosome', \n",
    "                        'position': 'base_pair_location', 'A1': 'effect_allele', \n",
    "                        'A2': 'other_allele', 'freq': 'effect_allele_frequency', \n",
    "                        'b': 'beta', 'se': 'standard_error', 'N_cases': 'n_cases', \n",
    "                        'N_controls': 'n_controls', 'REF': 'ref_allele', \n",
    "                        'ALT': 'alt_allele'}, inplace=True)\n",
    "# need to strip 'chr' from chromosome\n",
    "gwas_df.chromosome = gwas_df.chromosome.str.replace('chr', '')\n",
    "# make sure bp pos is int instead of float\n",
    "gwas_df.base_pair_location = gwas_df.base_pair_location.astype('int64')\n",
    "\n",
    "# reorder columns\n",
    "gwas_df = gwas_df[['variant_id', 'p_value', 'chromosome', 'base_pair_location', \n",
    "                   'effect_allele', 'other_allele', 'effect_allele_frequency', \n",
    "                   'beta', 'standard_error', 'n_cases', 'n_controls', \n",
    "                   'ref_allele', 'alt_allele']]\n",
    "\n",
    "print(f'gwas summary stats shape {gwas_df.shape}')\n",
    "if DEBUG:\n",
    "    display(gwas_df.head()) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "273f1210-2b32-4019-af27-213e44095c3b",
   "metadata": {},
   "source": [
    "#### compute the odds ratio values and add"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ac16502a-7ac8-417e-b6c6-f5ab82c66ccc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gwas summary stats shape (16293588, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>p_value</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>base_pair_location</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>effect_allele_frequency</th>\n",
       "      <th>beta</th>\n",
       "      <th>standard_error</th>\n",
       "      <th>n_cases</th>\n",
       "      <th>n_controls</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>odds_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs11020170</td>\n",
       "      <td>0.3771</td>\n",
       "      <td>11</td>\n",
       "      <td>88516209</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9931</td>\n",
       "      <td>0.1575</td>\n",
       "      <td>0.1783</td>\n",
       "      <td>7161</td>\n",
       "      <td>5356</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.170581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs116406626</td>\n",
       "      <td>0.1846</td>\n",
       "      <td>1</td>\n",
       "      <td>59855320</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9336</td>\n",
       "      <td>0.0605</td>\n",
       "      <td>0.0456</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.062368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs11992603</td>\n",
       "      <td>0.1887</td>\n",
       "      <td>8</td>\n",
       "      <td>134896404</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.2081</td>\n",
       "      <td>-0.0358</td>\n",
       "      <td>0.0273</td>\n",
       "      <td>26421</td>\n",
       "      <td>442271</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.964833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs192908256</td>\n",
       "      <td>0.8886</td>\n",
       "      <td>12</td>\n",
       "      <td>3762548</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9972</td>\n",
       "      <td>0.1489</td>\n",
       "      <td>1.0636</td>\n",
       "      <td>749</td>\n",
       "      <td>658</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>1.160557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs189372368</td>\n",
       "      <td>0.7543</td>\n",
       "      <td>16</td>\n",
       "      <td>77114961</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.9976</td>\n",
       "      <td>-0.1213</td>\n",
       "      <td>0.3874</td>\n",
       "      <td>6248</td>\n",
       "      <td>4391</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.885768</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    variant_id  p_value chromosome  base_pair_location effect_allele  \\\n",
       "0   rs11020170   0.3771         11            88516209             T   \n",
       "1  rs116406626   0.1846          1            59855320             A   \n",
       "2   rs11992603   0.1887          8           134896404             A   \n",
       "3  rs192908256   0.8886         12             3762548             A   \n",
       "4  rs189372368   0.7543         16            77114961             A   \n",
       "\n",
       "  other_allele  effect_allele_frequency    beta  standard_error  n_cases  \\\n",
       "0            C                   0.9931  0.1575          0.1783     7161   \n",
       "1            G                   0.9336  0.0605          0.0456    26421   \n",
       "2            G                   0.2081 -0.0358          0.0273    26421   \n",
       "3            C                   0.9972  0.1489          1.0636      749   \n",
       "4            G                   0.9976 -0.1213          0.3874     6248   \n",
       "\n",
       "   n_controls ref_allele alt_allele  odds_ratio  \n",
       "0        5356          T          C    1.170581  \n",
       "1      442271          A          G    1.062368  \n",
       "2      442271          G          A    0.964833  \n",
       "3         658          A          C    1.160557  \n",
       "4        4391          A          G    0.885768  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gwas_df['odds_ratio'] = exp(gwas_df.beta)\n",
    "\n",
    "print(f'gwas summary stats shape {gwas_df.shape}')\n",
    "if DEBUG:\n",
    "    display(gwas_df.head()) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "661a880c-23c9-4d48-b370-da7da50d3142",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>p_value</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>base_pair_location</th>\n",
       "      <th>effect_allele</th>\n",
       "      <th>other_allele</th>\n",
       "      <th>effect_allele_frequency</th>\n",
       "      <th>beta</th>\n",
       "      <th>standard_error</th>\n",
       "      <th>n_cases</th>\n",
       "      <th>n_controls</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>odds_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2764250</th>\n",
       "      <td>rs823118</td>\n",
       "      <td>4.941000e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>205754444</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.5748</td>\n",
       "      <td>0.0999</td>\n",
       "      <td>0.0171</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.105060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3539407</th>\n",
       "      <td>rs11557080</td>\n",
       "      <td>2.122000e-08</td>\n",
       "      <td>1</td>\n",
       "      <td>205768611</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.1426</td>\n",
       "      <td>0.1351</td>\n",
       "      <td>0.0241</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.144651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3893676</th>\n",
       "      <td>rs114138760</td>\n",
       "      <td>2.246000e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>154925709</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0112</td>\n",
       "      <td>0.3113</td>\n",
       "      <td>0.0844</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1.365199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6114374</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>5.764000e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>155235843</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.9930</td>\n",
       "      <td>-0.4907</td>\n",
       "      <td>0.1426</td>\n",
       "      <td>12989</td>\n",
       "      <td>10894</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.612198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6568248</th>\n",
       "      <td>rs76116224</td>\n",
       "      <td>1.190000e-04</td>\n",
       "      <td>2</td>\n",
       "      <td>17966582</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.9105</td>\n",
       "      <td>0.1551</td>\n",
       "      <td>0.0403</td>\n",
       "      <td>26948</td>\n",
       "      <td>442743</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1.167775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7573730</th>\n",
       "      <td>rs11578699</td>\n",
       "      <td>4.235000e-04</td>\n",
       "      <td>1</td>\n",
       "      <td>171750629</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.1955</td>\n",
       "      <td>-0.0781</td>\n",
       "      <td>0.0222</td>\n",
       "      <td>32505</td>\n",
       "      <td>448088</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.924872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9698034</th>\n",
       "      <td>rs6658353</td>\n",
       "      <td>2.418000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>161499264</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>0.5014</td>\n",
       "      <td>0.0722</td>\n",
       "      <td>0.0171</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1.074870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10185847</th>\n",
       "      <td>rs4653767</td>\n",
       "      <td>8.667000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>226728377</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.7158</td>\n",
       "      <td>0.0729</td>\n",
       "      <td>0.0186</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1.075623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13526195</th>\n",
       "      <td>rs35749011</td>\n",
       "      <td>5.022000e-30</td>\n",
       "      <td>1</td>\n",
       "      <td>155162560</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.0191</td>\n",
       "      <td>0.7508</td>\n",
       "      <td>0.0659</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2.118694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15191523</th>\n",
       "      <td>rs10797576</td>\n",
       "      <td>3.532000e-05</td>\n",
       "      <td>1</td>\n",
       "      <td>232528865</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.1434</td>\n",
       "      <td>0.0998</td>\n",
       "      <td>0.0241</td>\n",
       "      <td>33674</td>\n",
       "      <td>449056</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.104950</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           variant_id       p_value chromosome  base_pair_location  \\\n",
       "2764250      rs823118  4.941000e-09          1           205754444   \n",
       "3539407    rs11557080  2.122000e-08          1           205768611   \n",
       "3893676   rs114138760  2.246000e-04          1           154925709   \n",
       "6114374    rs76763715  5.764000e-04          1           155235843   \n",
       "6568248    rs76116224  1.190000e-04          2            17966582   \n",
       "7573730    rs11578699  4.235000e-04          1           171750629   \n",
       "9698034     rs6658353  2.418000e-05          1           161499264   \n",
       "10185847    rs4653767  8.667000e-05          1           226728377   \n",
       "13526195   rs35749011  5.022000e-30          1           155162560   \n",
       "15191523   rs10797576  3.532000e-05          1           232528865   \n",
       "\n",
       "         effect_allele other_allele  effect_allele_frequency    beta  \\\n",
       "2764250              T            C                   0.5748  0.0999   \n",
       "3539407              A            G                   0.1426  0.1351   \n",
       "3893676              C            G                   0.0112  0.3113   \n",
       "6114374              T            C                   0.9930 -0.4907   \n",
       "6568248              A            T                   0.9105  0.1551   \n",
       "7573730              T            C                   0.1955 -0.0781   \n",
       "9698034              C            G                   0.5014  0.0722   \n",
       "10185847             T            C                   0.7158  0.0729   \n",
       "13526195             A            G                   0.0191  0.7508   \n",
       "15191523             T            C                   0.1434  0.0998   \n",
       "\n",
       "          standard_error  n_cases  n_controls ref_allele alt_allele  \\\n",
       "2764250           0.0171    33674      449056          C          T   \n",
       "3539407           0.0241    33674      449056          G          A   \n",
       "3893676           0.0844    33674      449056          G          C   \n",
       "6114374           0.1426    12989       10894          T          C   \n",
       "6568248           0.0403    26948      442743          A          T   \n",
       "7573730           0.0222    32505      448088          C          T   \n",
       "9698034           0.0171    33674      449056          G          C   \n",
       "10185847          0.0186    33674      449056          T          C   \n",
       "13526195          0.0659    33674      449056          G          A   \n",
       "15191523          0.0241    33674      449056          C          T   \n",
       "\n",
       "          odds_ratio  \n",
       "2764250     1.105060  \n",
       "3539407     1.144651  \n",
       "3893676     1.365199  \n",
       "6114374     0.612198  \n",
       "6568248     1.167775  \n",
       "7573730     0.924872  \n",
       "9698034     1.074870  \n",
       "10185847    1.075623  \n",
       "13526195    2.118694  \n",
       "15191523    1.104950  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(gwas_df.loc[gwas_df.variant_id.isin(test_snps)])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f944cff-b3d4-459f-81ed-8ffee5d8c45f",
   "metadata": {},
   "source": [
    "### save the formatted and annotated GWAS summary statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a13b1759-6fb6-4121-854b-b916f99f6f49",
   "metadata": {},
   "source": [
    "#### save in file format to match NHGRI-EBI GWAS Catalog"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2e059846-fb2e-49da-a07a-ecbbae164269",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_df.to_csv(out_catalogue_file, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4215c43e-bebf-49a5-a7ec-1fd0176d73ec",
   "metadata": {},
   "source": [
    "#### save in parquet for faster access for other projects, etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a8dd5129-2460-44f2-a464-c0178b79eca8",
   "metadata": {},
   "outputs": [],
   "source": [
    "gwas_df.to_parquet(out_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "42820cfa-bbf1-4f1d-a7ff-89bfd6444ebc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Jul 29 17:12:43 UTC 2022\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "python3",
   "name": "pytorch-gpu.1-11.m94",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/pytorch-gpu.1-11:m94"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
