{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Notebook to run *risk*-eQTL analysis using [tensorQTL](https://github.com/broadinstitute/tensorqtl)\n",
    "\n",
    "ie all Dx risk index variants against all transcripts so not *cis* or *trans* \n",
    "\n",
    "[Taylor-Weiner, Aguet, et al., Genome Biol. 20:228, 2019.](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1836-7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Feb 19 05:23:58 UTC 2021\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### import libraries and set notebook variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: 'rfunc' cannot be imported. R and the 'rpy2' Python package are needed.\n",
      "PyTorch 1.7.1\n",
      "Pandas 1.2.2\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import torch\n",
    "import tensorqtl.tensorqtl as tensorqtl\n",
    "from tensorqtl.tensorqtl import genotypeio, cis, trans\n",
    "print('PyTorch {}'.format(torch.__version__))\n",
    "print('Pandas {}'.format(pd.__version__))\n",
    "\n",
    "import os\n",
    "import numpy as np\n",
    "import statsmodels.stats.multitest as smm\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# import random\n",
    "# import threading\n",
    "# import dask.dataframe as dd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# parameters\n",
    "cohort = 'foundin'\n",
    "amp_abbr = 'PP'\n",
    "version = 'amppdv1'\n",
    "day = 'da65'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# naming\n",
    "cohort_version = f'{cohort}.{version}'\n",
    "cohort_build = f'{cohort}.{day}'\n",
    "\n",
    "# directories\n",
    "wrk_dir = f'/home/jupyter/{cohort}/eqtl'\n",
    "geno_dir = f'{wrk_dir}/genotypes'\n",
    "expr_dir = f'{wrk_dir}/expression'\n",
    "info_dir = f'{wrk_dir}/sample_info'\n",
    "tensorqtl_dir = f'{wrk_dir}/tensorqtl'\n",
    "results_dir = f'{wrk_dir}/results'\n",
    "\n",
    "# input files\n",
    "expr_bed_file = f'{expr_dir}/{cohort_build}.norm.adj.bed.gz'\n",
    "bfile_prefix_path = f'{geno_dir}/{cohort_version}.risk.bfile'\n",
    "assay_covs_files = f'{info_dir}/foundin_rnab_sample_info.csv'\n",
    "# gencode_pkl = f'{expr_dir}/gencode_v29.lncipedia_v5_2_hc.annotation.pkl'\n",
    "# pfiles = '{genodir}/{cohortversion}.chr{chr}'\n",
    "# genome_index_file = f'{expr_dir}/Homo_sapiens_assembly38.fasta.fai'\n",
    "\n",
    "# output files\n",
    "# eqtl_psam = f'{info_dir}/{cohort_build}.psam'\n",
    "# chr_detected_out_file = '{exprdir}/{cohortbuild}.detected.genes.chr{chr}'\n",
    "# used_samples_list_file = f'{info_dir}/{cohort_version}.eqtl.samples'\n",
    "risk_results_file = f'{tensorqtl_dir}/{cohort_build}.risk_qtl_pairs.parquet'\n",
    "risk_cs_results_file = f'{tensorqtl_dir}/{cohort_build}.cs.risk_qtl_pairs.parquet'\n",
    "sig_risk_cs_results_file = f'{results_dir}/{cohort_build}.cs.risk_qtl_pairs.csv'\n",
    "\n",
    "# constant values\n",
    "# autosomes = [str(x) for x in list(range(1,23))]\n",
    "# max_dist = 1000000\n",
    "# capture_out = !(nproc)\n",
    "# max_threads = int(capture_out[0])\n",
    "alpha_value = 0.05\n",
    "min_nominal_alpha = 1e-05\n",
    "# max_feature_cnt_parallel_load = 20000\n",
    "\n",
    "cell_types_list = ['DopaminergicNeurons', 'EarlyneuronProgenitor',\n",
    "                   'ImmatureDopaminergicNeurons', 'ProliferatingFloorPlateProgenitors',\n",
    "                   'LateneuronProgenitor', 'Ependymal-likeCells',\n",
    "                   'Neuroepithelial-likeCells']\n",
    "repeated_samples_to_exclude = ['RNAB_PPMI3966B1_2813_da65_v1', 'RNAB_PPMI3966B2_2813_da65_v2',\n",
    "                               'RNAB_PPMI3966B5_2813_da65_v1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(tensorqtl_dir, exist_ok=True)\n",
    "os.makedirs(results_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### utility functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute B&H FDR for given p-values\n",
    "def compute_fdr(pvalues):\n",
    "    bh_adj = smm.fdrcorrection(pvalues)\n",
    "    return bh_adj[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### load phenotypes and covariates (if needed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "\n",
    "phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expr_bed_file)\n",
    "covs_df = pd.read_csv(assay_covs_files, index_col=0)\n",
    "print(f'phenotype_df {phenotype_df.shape}')\n",
    "print(f'phenotype_pos_df {phenotype_df.shape}')\n",
    "print(f'covariates_df {phenotype_df.shape}')\n",
    "# display(phenotype_df.head())\n",
    "# display(phenotype_pos_df.head())\n",
    "# display(covs_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### load plink bfiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:00<00:00, 160.51it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 44.8 ms, sys: 1.12 ms, total: 45.9 ms\n",
      "Wall time: 57.9 ms\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# pr = genotypeio.PlinkReader(bfile_prefix_path, select_samples=phenotype_df.columns)\n",
    "pr = genotypeio.PlinkReader(bfile_prefix_path)\n",
    "genotype_df = pr.load_genotypes()\n",
    "variant_df = pr.bim.set_index('snp')[['chrom', 'pos']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(genotype_df.shape)\n",
    "# display(genotype_df.head())\n",
    "print(variant_df.shape)\n",
    "# display(variant_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(104, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snp</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs76763715</th>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs6658353</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161499264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs11578699</th>\n",
       "      <td>chr1</td>\n",
       "      <td>171750629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs6676110</th>\n",
       "      <td>chr1</td>\n",
       "      <td>205693849</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs823118</th>\n",
       "      <td>chr1</td>\n",
       "      <td>205754444</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           chrom        pos\n",
       "snp                        \n",
       "rs76763715  chr1  155235843\n",
       "rs6658353   chr1  161499264\n",
       "rs11578699  chr1  171750629\n",
       "rs6676110   chr1  205693849\n",
       "rs823118    chr1  205754444"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# tensorQTL says wants plink bfiles, but wants bim chrs to include 'chr'\n",
    "variant_df['chrom'] = 'chr' + variant_df['chrom']\n",
    "print(variant_df.shape)\n",
    "display(variant_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### make sure the pheno and genos have same samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intersect 91\n",
      "number of genotypes samples not in expression 28\n",
      "number of expression samples not in genotypes 28\n"
     ]
    }
   ],
   "source": [
    "assay_intersect_samples = set(genotype_df.columns) & set(phenotype_df.columns) \n",
    "print(f'intersect {len(assay_intersect_samples)}')\n",
    "extra_geno_samples = set(genotype_df.columns) - set(phenotype_df.columns)\n",
    "print(f'number of genotypes samples not in expression {len(extra_geno_samples)}')\n",
    "extra_expr_samples = set(phenotype_df.columns) - set(genotype_df.columns)\n",
    "print(f'number of expression samples not in genotypes {len(extra_geno_samples)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### drop the non-matched samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotype_df.drop(columns=extra_geno_samples, inplace=True)\n",
    "phenotype_df.drop(columns=extra_expr_samples, inplace=True)\n",
    "\n",
    "print(genotype_df.shape)\n",
    "# display(genotype_df.head())\n",
    "print(phenotype_df.shape)\n",
    "# display(phenotype_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### need to make sure phenos and genos have matched chromosomes; ie just autosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intersect 21\n",
      "number of genotypes chroms not in expression 0\n",
      "set()\n",
      "number of expression chroms not in genotypes 12\n",
      "{'KI270713.1', 'GL000195.1', 'chrX', 'chr22', 'KI270733.1', 'KI270711.1', 'KI270728.1', 'GL000219.1', 'KI270734.1', 'chrM', 'KI270727.1', 'chrY'}\n"
     ]
    }
   ],
   "source": [
    "# need to ditch any non-autosomal genes\n",
    "assay_intersect_chroms = set(phenotype_pos_df['chr']) & set(variant_df['chrom']) \n",
    "print(f'intersect {len(assay_intersect_chroms)}')\n",
    "extra_geno_chroms = set(variant_df['chrom']) - set(phenotype_pos_df['chr'])\n",
    "print(f'number of genotypes chroms not in expression {len(extra_geno_chroms)}')\n",
    "print(extra_geno_chroms)\n",
    "extra_expr_chroms = set(phenotype_pos_df['chr']) - set(variant_df['chrom'])\n",
    "print(f'number of expression chroms not in genotypes {len(extra_expr_chroms)}')\n",
    "print(extra_expr_chroms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(extra_geno_chroms) > 0:\n",
    "    variant_df = variant_df.loc[~variant_df['chrom'].isin(extra_geno_chroms)]\n",
    "    # this will remove variants so need to remove them from genos df as well\n",
    "    genotype_df = genotype_df.loc[genotype_df.index.isin(variant_df.index)]\n",
    "if len(extra_expr_chroms) > 0:\n",
    "    phenotype_pos_df = phenotype_pos_df.loc[~phenotype_pos_df['chr'].isin(extra_expr_chroms)]\n",
    "    # this will remove genes so need to remove them from phenos df as well\n",
    "    phenotype_df = phenotype_df.loc[phenotype_df.index.isin(phenotype_pos_df.index)]\n",
    "\n",
    "print(genotype_df.shape)\n",
    "# display(genotype_df.head())\n",
    "print(variant_df.shape)\n",
    "# display(variant_df.head())\n",
    "print(phenotype_df.shape)\n",
    "# display(phenotype_df.head())\n",
    "print(phenotype_pos_df.shape)\n",
    "# display(phenotype_pos_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### make sure covariates match geno and pheno samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(91, 98)\n",
      "intersect 91\n",
      "number of expression samples not in covariates 0\n",
      "number of covariate samples not in genotypes 0\n"
     ]
    }
   ],
   "source": [
    "# subest covs to just this 'day'; ie all differention days covs in file\n",
    "# also since only interested in cell fractions as interaction terms, subset now\n",
    "covs_df = covs_df.loc[(covs_df['day'] == day) & (covs_df['sampleid'].isin(phenotype_df.columns))]\n",
    "covs_df = covs_df.loc[~covs_df.index.isin(repeated_samples_to_exclude)]\n",
    "print(covs_df.shape)\n",
    "\n",
    "cov_intersect_samples = set(phenotype_df.columns) & set(covs_df['sampleid']) \n",
    "print(f'intersect {len(cov_intersect_samples)}')\n",
    "extra_expr_samples = set(phenotype_df.columns) - set(covs_df['sampleid'])\n",
    "print(f'number of expression samples not in covariates {len(extra_expr_samples)}')\n",
    "extra_cov_samples = set(covs_df['sampleid']) - set(phenotype_df.columns)\n",
    "print(f'number of covariate samples not in genotypes {len(extra_cov_samples)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### *risk*-QTL: nominal p-values for all risk variant-phenotype pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trans-QTL mapping\n",
      "  * 91 samples\n",
      "  * 30944 phenotypes\n",
      "  * 104 variants\n",
      "    processing batch 1/1\n",
      "    elapsed time: 0.02 min\n",
      "  * 104 variants passed MAF >= 0.01 filtering\n",
      "done.\n",
      "(3218176, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>b</th>\n",
       "      <th>b_se</th>\n",
       "      <th>r2</th>\n",
       "      <th>maf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000116191.17</td>\n",
       "      <td>0.914136</td>\n",
       "      <td>-0.028271</td>\n",
       "      <td>0.261459</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.10989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000138772.12</td>\n",
       "      <td>0.326212</td>\n",
       "      <td>0.253483</td>\n",
       "      <td>0.256765</td>\n",
       "      <td>0.010832</td>\n",
       "      <td>0.10989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>lnc_RPL18_3</td>\n",
       "      <td>0.255686</td>\n",
       "      <td>-0.274059</td>\n",
       "      <td>0.239560</td>\n",
       "      <td>0.014492</td>\n",
       "      <td>0.10989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>lnc_SLC16A4_3</td>\n",
       "      <td>0.976111</td>\n",
       "      <td>0.007849</td>\n",
       "      <td>0.261382</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.10989</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000137709.9</td>\n",
       "      <td>0.899067</td>\n",
       "      <td>-0.032036</td>\n",
       "      <td>0.251854</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>0.10989</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   variant_id        phenotype_id      pval         b      b_se        r2  \\\n",
       "0  rs76763715  ENSG00000116191.17  0.914136 -0.028271  0.261459  0.000131   \n",
       "1  rs76763715  ENSG00000138772.12  0.326212  0.253483  0.256765  0.010832   \n",
       "2  rs76763715         lnc_RPL18_3  0.255686 -0.274059  0.239560  0.014492   \n",
       "3  rs76763715       lnc_SLC16A4_3  0.976111  0.007849  0.261382  0.000010   \n",
       "4  rs76763715   ENSG00000137709.9  0.899067 -0.032036  0.251854  0.000182   \n",
       "\n",
       "       maf  \n",
       "0  0.10989  \n",
       "1  0.10989  \n",
       "2  0.10989  \n",
       "3  0.10989  \n",
       "4  0.10989  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 6.46 s, sys: 1.53 s, total: 7.99 s\n",
      "Wall time: 6.59 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# run mapping\n",
    "trans_df = trans.map_trans(genotype_df, phenotype_df, batch_size=10000, return_r2=True,\n",
    "                           return_sparse=True, pval_threshold=1, maf_threshold=0.01)\n",
    "print(trans_df.shape)\n",
    "display(trans_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### compute the FDR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the corrected p-value, note just based on all chrom gene pvalues    \n",
    "trans_df['bh_fdr'] = compute_fdr(trans_df['pval'].fillna(1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3218176, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>b</th>\n",
       "      <th>b_se</th>\n",
       "      <th>r2</th>\n",
       "      <th>maf</th>\n",
       "      <th>bh_fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000116191.17</td>\n",
       "      <td>0.914136</td>\n",
       "      <td>-0.028271</td>\n",
       "      <td>0.261459</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.999955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000138772.12</td>\n",
       "      <td>0.326212</td>\n",
       "      <td>0.253483</td>\n",
       "      <td>0.256765</td>\n",
       "      <td>0.010832</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.999632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>lnc_RPL18_3</td>\n",
       "      <td>0.255686</td>\n",
       "      <td>-0.274059</td>\n",
       "      <td>0.239560</td>\n",
       "      <td>0.014492</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.998207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>lnc_SLC16A4_3</td>\n",
       "      <td>0.976111</td>\n",
       "      <td>0.007849</td>\n",
       "      <td>0.261382</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.999998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000137709.9</td>\n",
       "      <td>0.899067</td>\n",
       "      <td>-0.032036</td>\n",
       "      <td>0.251854</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.999955</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   variant_id        phenotype_id      pval         b      b_se        r2  \\\n",
       "0  rs76763715  ENSG00000116191.17  0.914136 -0.028271  0.261459  0.000131   \n",
       "1  rs76763715  ENSG00000138772.12  0.326212  0.253483  0.256765  0.010832   \n",
       "2  rs76763715         lnc_RPL18_3  0.255686 -0.274059  0.239560  0.014492   \n",
       "3  rs76763715       lnc_SLC16A4_3  0.976111  0.007849  0.261382  0.000010   \n",
       "4  rs76763715   ENSG00000137709.9  0.899067 -0.032036  0.251854  0.000182   \n",
       "\n",
       "       maf    bh_fdr  \n",
       "0  0.10989  0.999955  \n",
       "1  0.10989  0.999632  \n",
       "2  0.10989  0.998207  \n",
       "3  0.10989  0.999998  \n",
       "4  0.10989  0.999955  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(trans_df.shape)\n",
    "display(trans_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    3.218176e+06\n",
       "mean     9.955343e-01\n",
       "std      1.445357e-02\n",
       "min      1.445746e-05\n",
       "25%      9.980010e-01\n",
       "50%      9.999552e-01\n",
       "75%      9.999552e-01\n",
       "max      1.000000e+00\n",
       "Name: bh_fdr, dtype: float64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_df['bh_fdr'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3218176, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>b</th>\n",
       "      <th>b_se</th>\n",
       "      <th>r2</th>\n",
       "      <th>maf</th>\n",
       "      <th>bh_fdr</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>chr</th>\n",
       "      <th>tss</th>\n",
       "      <th>onchrom</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000116191.17</td>\n",
       "      <td>0.914136</td>\n",
       "      <td>-0.028271</td>\n",
       "      <td>0.261459</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.999955</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843</td>\n",
       "      <td>chr1</td>\n",
       "      <td>178725147</td>\n",
       "      <td>on</td>\n",
       "      <td>23489304.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000138772.12</td>\n",
       "      <td>0.326212</td>\n",
       "      <td>0.253483</td>\n",
       "      <td>0.256765</td>\n",
       "      <td>0.010832</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.999632</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843</td>\n",
       "      <td>chr4</td>\n",
       "      <td>78551519</td>\n",
       "      <td>between</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>lnc_RPL18_3</td>\n",
       "      <td>0.255686</td>\n",
       "      <td>-0.274059</td>\n",
       "      <td>0.239560</td>\n",
       "      <td>0.014492</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.998207</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843</td>\n",
       "      <td>chr19</td>\n",
       "      <td>48624132</td>\n",
       "      <td>between</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>lnc_SLC16A4_3</td>\n",
       "      <td>0.976111</td>\n",
       "      <td>0.007849</td>\n",
       "      <td>0.261382</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.999998</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843</td>\n",
       "      <td>chr1</td>\n",
       "      <td>110338737</td>\n",
       "      <td>on</td>\n",
       "      <td>-44897106.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000137709.9</td>\n",
       "      <td>0.899067</td>\n",
       "      <td>-0.032036</td>\n",
       "      <td>0.251854</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.999955</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843</td>\n",
       "      <td>chr11</td>\n",
       "      <td>120236640</td>\n",
       "      <td>between</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   variant_id        phenotype_id      pval         b      b_se        r2  \\\n",
       "0  rs76763715  ENSG00000116191.17  0.914136 -0.028271  0.261459  0.000131   \n",
       "1  rs76763715  ENSG00000138772.12  0.326212  0.253483  0.256765  0.010832   \n",
       "2  rs76763715         lnc_RPL18_3  0.255686 -0.274059  0.239560  0.014492   \n",
       "3  rs76763715       lnc_SLC16A4_3  0.976111  0.007849  0.261382  0.000010   \n",
       "4  rs76763715   ENSG00000137709.9  0.899067 -0.032036  0.251854  0.000182   \n",
       "\n",
       "       maf    bh_fdr chrom        pos    chr        tss  onchrom        dist  \n",
       "0  0.10989  0.999955  chr1  155235843   chr1  178725147       on  23489304.0  \n",
       "1  0.10989  0.999632  chr1  155235843   chr4   78551519  between         NaN  \n",
       "2  0.10989  0.998207  chr1  155235843  chr19   48624132  between         NaN  \n",
       "3  0.10989  0.999998  chr1  155235843   chr1  110338737       on -44897106.0  \n",
       "4  0.10989  0.999955  chr1  155235843  chr11  120236640  between         NaN  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# add physical positions onto results\n",
    "trans_df = trans_df.merge(variant_df, how='left', \n",
    "                          left_on='variant_id', right_on='snp')\n",
    "trans_df = trans_df.merge(phenotype_pos_df, how='left', \n",
    "                          left_on='phenotype_id', right_on='phenotype_id')\n",
    "trans_df['onchrom'] = np.where(trans_df['chrom'] == trans_df['chr'], \n",
    "                               'on', 'between')\n",
    "trans_df['dist'] = trans_df['tss'] - trans_df['pos']\n",
    "trans_df.loc[trans_df['onchrom'] == 'between', 'dist'] = np.nan\n",
    "\n",
    "print(trans_df.shape)\n",
    "display(trans_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(21, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>b</th>\n",
       "      <th>b_se</th>\n",
       "      <th>r2</th>\n",
       "      <th>maf</th>\n",
       "      <th>bh_fdr</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>chr</th>\n",
       "      <th>tss</th>\n",
       "      <th>onchrom</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>138444</th>\n",
       "      <td>rs823118</td>\n",
       "      <td>lnc_RHEX_8</td>\n",
       "      <td>4.879842e-08</td>\n",
       "      <td>-0.798306</td>\n",
       "      <td>0.133847</td>\n",
       "      <td>0.285561</td>\n",
       "      <td>0.362637</td>\n",
       "      <td>0.008725</td>\n",
       "      <td>chr1</td>\n",
       "      <td>205754444</td>\n",
       "      <td>chr1</td>\n",
       "      <td>205862522</td>\n",
       "      <td>on</td>\n",
       "      <td>108078.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417630</th>\n",
       "      <td>rs73038319</td>\n",
       "      <td>lnc_C1QTNF3_AMACR_3</td>\n",
       "      <td>4.492441e-12</td>\n",
       "      <td>-3.231971</td>\n",
       "      <td>0.404354</td>\n",
       "      <td>0.417870</td>\n",
       "      <td>0.021978</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>chr3</td>\n",
       "      <td>18320267</td>\n",
       "      <td>chr5</td>\n",
       "      <td>34190204</td>\n",
       "      <td>between</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103755</th>\n",
       "      <td>rs26431</td>\n",
       "      <td>ENSG00000175749.11</td>\n",
       "      <td>1.059660e-07</td>\n",
       "      <td>0.840778</td>\n",
       "      <td>0.145314</td>\n",
       "      <td>0.273335</td>\n",
       "      <td>0.241758</td>\n",
       "      <td>0.017051</td>\n",
       "      <td>chr5</td>\n",
       "      <td>103030090</td>\n",
       "      <td>chr5</td>\n",
       "      <td>103032376</td>\n",
       "      <td>on</td>\n",
       "      <td>2286.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1415204</th>\n",
       "      <td>rs199351</td>\n",
       "      <td>KLHL7_DT</td>\n",
       "      <td>6.943582e-09</td>\n",
       "      <td>0.816349</td>\n",
       "      <td>0.127466</td>\n",
       "      <td>0.315474</td>\n",
       "      <td>0.445055</td>\n",
       "      <td>0.001490</td>\n",
       "      <td>chr7</td>\n",
       "      <td>23260430</td>\n",
       "      <td>chr7</td>\n",
       "      <td>23105703</td>\n",
       "      <td>on</td>\n",
       "      <td>-154727.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2786097</th>\n",
       "      <td>rs142167</td>\n",
       "      <td>lnc_PLEKHM1_3</td>\n",
       "      <td>1.082803e-10</td>\n",
       "      <td>-1.052603</td>\n",
       "      <td>0.143912</td>\n",
       "      <td>0.375429</td>\n",
       "      <td>0.296703</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46717868</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45520464</td>\n",
       "      <td>on</td>\n",
       "      <td>-1197404.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         variant_id         phenotype_id          pval         b      b_se  \\\n",
       "138444     rs823118           lnc_RHEX_8  4.879842e-08 -0.798306  0.133847   \n",
       "417630   rs73038319  lnc_C1QTNF3_AMACR_3  4.492441e-12 -3.231971  0.404354   \n",
       "1103755     rs26431   ENSG00000175749.11  1.059660e-07  0.840778  0.145314   \n",
       "1415204    rs199351             KLHL7_DT  6.943582e-09  0.816349  0.127466   \n",
       "2786097    rs142167        lnc_PLEKHM1_3  1.082803e-10 -1.052603  0.143912   \n",
       "\n",
       "               r2       maf    bh_fdr  chrom        pos    chr        tss  \\\n",
       "138444   0.285561  0.362637  0.008725   chr1  205754444   chr1  205862522   \n",
       "417630   0.417870  0.021978  0.000014   chr3   18320267   chr5   34190204   \n",
       "1103755  0.273335  0.241758  0.017051   chr5  103030090   chr5  103032376   \n",
       "1415204  0.315474  0.445055  0.001490   chr7   23260430   chr7   23105703   \n",
       "2786097  0.375429  0.296703  0.000058  chr17   46717868  chr17   45520464   \n",
       "\n",
       "         onchrom       dist  \n",
       "138444        on   108078.0  \n",
       "417630   between        NaN  \n",
       "1103755       on     2286.0  \n",
       "1415204       on  -154727.0  \n",
       "2786097       on -1197404.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sig_risk_df = trans_df.loc[trans_df['bh_fdr'] <= alpha_value]\n",
    "print(sig_risk_df.shape)\n",
    "display(sig_risk_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval</th>\n",
       "      <th>b</th>\n",
       "      <th>b_se</th>\n",
       "      <th>r2</th>\n",
       "      <th>maf</th>\n",
       "      <th>bh_fdr</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>chr</th>\n",
       "      <th>tss</th>\n",
       "      <th>onchrom</th>\n",
       "      <th>dist</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>138444</th>\n",
       "      <td>rs823118</td>\n",
       "      <td>lnc_RHEX_8</td>\n",
       "      <td>4.879842e-08</td>\n",
       "      <td>-0.798306</td>\n",
       "      <td>0.133847</td>\n",
       "      <td>0.285561</td>\n",
       "      <td>0.362637</td>\n",
       "      <td>0.008725</td>\n",
       "      <td>chr1</td>\n",
       "      <td>205754444</td>\n",
       "      <td>chr1</td>\n",
       "      <td>205862522</td>\n",
       "      <td>on</td>\n",
       "      <td>108078.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>417630</th>\n",
       "      <td>rs73038319</td>\n",
       "      <td>lnc_C1QTNF3_AMACR_3</td>\n",
       "      <td>4.492441e-12</td>\n",
       "      <td>-3.231971</td>\n",
       "      <td>0.404354</td>\n",
       "      <td>0.417870</td>\n",
       "      <td>0.021978</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>chr3</td>\n",
       "      <td>18320267</td>\n",
       "      <td>chr5</td>\n",
       "      <td>34190204</td>\n",
       "      <td>between</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103755</th>\n",
       "      <td>rs26431</td>\n",
       "      <td>ENSG00000175749.11</td>\n",
       "      <td>1.059660e-07</td>\n",
       "      <td>0.840778</td>\n",
       "      <td>0.145314</td>\n",
       "      <td>0.273335</td>\n",
       "      <td>0.241758</td>\n",
       "      <td>0.017051</td>\n",
       "      <td>chr5</td>\n",
       "      <td>103030090</td>\n",
       "      <td>chr5</td>\n",
       "      <td>103032376</td>\n",
       "      <td>on</td>\n",
       "      <td>2286.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1415204</th>\n",
       "      <td>rs199351</td>\n",
       "      <td>KLHL7_DT</td>\n",
       "      <td>6.943582e-09</td>\n",
       "      <td>0.816349</td>\n",
       "      <td>0.127466</td>\n",
       "      <td>0.315474</td>\n",
       "      <td>0.445055</td>\n",
       "      <td>0.001490</td>\n",
       "      <td>chr7</td>\n",
       "      <td>23260430</td>\n",
       "      <td>chr7</td>\n",
       "      <td>23105703</td>\n",
       "      <td>on</td>\n",
       "      <td>-154727.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2786097</th>\n",
       "      <td>rs142167</td>\n",
       "      <td>lnc_PLEKHM1_3</td>\n",
       "      <td>1.082803e-10</td>\n",
       "      <td>-1.052603</td>\n",
       "      <td>0.143912</td>\n",
       "      <td>0.375429</td>\n",
       "      <td>0.296703</td>\n",
       "      <td>0.000058</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46717868</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45520464</td>\n",
       "      <td>on</td>\n",
       "      <td>-1197404.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2796934</th>\n",
       "      <td>rs142167</td>\n",
       "      <td>lnc_PLEKHM1_4</td>\n",
       "      <td>3.646967e-09</td>\n",
       "      <td>-0.957676</td>\n",
       "      <td>0.146265</td>\n",
       "      <td>0.325094</td>\n",
       "      <td>0.296703</td>\n",
       "      <td>0.000978</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46717868</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45516128</td>\n",
       "      <td>on</td>\n",
       "      <td>-1201740.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2801832</th>\n",
       "      <td>rs142167</td>\n",
       "      <td>lnc_LINC02210_CRHR1_7</td>\n",
       "      <td>1.056680e-09</td>\n",
       "      <td>-1.011352</td>\n",
       "      <td>0.148288</td>\n",
       "      <td>0.343246</td>\n",
       "      <td>0.296703</td>\n",
       "      <td>0.000340</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46717868</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45585004</td>\n",
       "      <td>on</td>\n",
       "      <td>-1132864.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2807509</th>\n",
       "      <td>rs142167</td>\n",
       "      <td>KANSL1_AS1</td>\n",
       "      <td>1.697743e-07</td>\n",
       "      <td>0.887911</td>\n",
       "      <td>0.156420</td>\n",
       "      <td>0.265811</td>\n",
       "      <td>0.296703</td>\n",
       "      <td>0.026017</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46717868</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46193573</td>\n",
       "      <td>on</td>\n",
       "      <td>-524295.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2809678</th>\n",
       "      <td>rs142167</td>\n",
       "      <td>ENSG00000238083.7</td>\n",
       "      <td>9.447630e-12</td>\n",
       "      <td>1.094144</td>\n",
       "      <td>0.139643</td>\n",
       "      <td>0.408212</td>\n",
       "      <td>0.296703</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46717868</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46511511</td>\n",
       "      <td>on</td>\n",
       "      <td>-206357.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2809708</th>\n",
       "      <td>rs142167</td>\n",
       "      <td>ENSG00000214425.7</td>\n",
       "      <td>8.185168e-11</td>\n",
       "      <td>-1.053283</td>\n",
       "      <td>0.142831</td>\n",
       "      <td>0.379276</td>\n",
       "      <td>0.296703</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46717868</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45550335</td>\n",
       "      <td>on</td>\n",
       "      <td>-1167533.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2878929</th>\n",
       "      <td>rs199502</td>\n",
       "      <td>lnc_PLEKHM1_3</td>\n",
       "      <td>2.304677e-09</td>\n",
       "      <td>-0.979609</td>\n",
       "      <td>0.147333</td>\n",
       "      <td>0.331874</td>\n",
       "      <td>0.269231</td>\n",
       "      <td>0.000674</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46784981</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45520464</td>\n",
       "      <td>on</td>\n",
       "      <td>-1264517.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2889766</th>\n",
       "      <td>rs199502</td>\n",
       "      <td>lnc_PLEKHM1_4</td>\n",
       "      <td>9.029755e-09</td>\n",
       "      <td>-0.927934</td>\n",
       "      <td>0.146229</td>\n",
       "      <td>0.311513</td>\n",
       "      <td>0.269231</td>\n",
       "      <td>0.001816</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46784981</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45516128</td>\n",
       "      <td>on</td>\n",
       "      <td>-1268853.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2894664</th>\n",
       "      <td>rs199502</td>\n",
       "      <td>lnc_LINC02210_CRHR1_7</td>\n",
       "      <td>2.025256e-10</td>\n",
       "      <td>-1.034765</td>\n",
       "      <td>0.144133</td>\n",
       "      <td>0.366736</td>\n",
       "      <td>0.269231</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46784981</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45585004</td>\n",
       "      <td>on</td>\n",
       "      <td>-1199977.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2900341</th>\n",
       "      <td>rs199502</td>\n",
       "      <td>KANSL1_AS1</td>\n",
       "      <td>3.307256e-08</td>\n",
       "      <td>0.920575</td>\n",
       "      <td>0.152085</td>\n",
       "      <td>0.291623</td>\n",
       "      <td>0.269231</td>\n",
       "      <td>0.006261</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46784981</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46193573</td>\n",
       "      <td>on</td>\n",
       "      <td>-591408.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2902510</th>\n",
       "      <td>rs199502</td>\n",
       "      <td>ENSG00000238083.7</td>\n",
       "      <td>8.092927e-11</td>\n",
       "      <td>1.044152</td>\n",
       "      <td>0.141546</td>\n",
       "      <td>0.379432</td>\n",
       "      <td>0.269231</td>\n",
       "      <td>0.000053</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46784981</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46511511</td>\n",
       "      <td>on</td>\n",
       "      <td>-273470.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2902540</th>\n",
       "      <td>rs199502</td>\n",
       "      <td>ENSG00000214425.7</td>\n",
       "      <td>4.550004e-11</td>\n",
       "      <td>-1.053522</td>\n",
       "      <td>0.140465</td>\n",
       "      <td>0.387277</td>\n",
       "      <td>0.269231</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46784981</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45550335</td>\n",
       "      <td>on</td>\n",
       "      <td>-1234646.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2909873</th>\n",
       "      <td>rs2074404</td>\n",
       "      <td>lnc_PLEKHM1_3</td>\n",
       "      <td>4.584136e-10</td>\n",
       "      <td>-0.859158</td>\n",
       "      <td>0.122697</td>\n",
       "      <td>0.355222</td>\n",
       "      <td>0.346154</td>\n",
       "      <td>0.000164</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46788073</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45520464</td>\n",
       "      <td>on</td>\n",
       "      <td>-1267609.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2920710</th>\n",
       "      <td>rs2074404</td>\n",
       "      <td>lnc_PLEKHM1_4</td>\n",
       "      <td>4.108276e-09</td>\n",
       "      <td>-0.801411</td>\n",
       "      <td>0.122894</td>\n",
       "      <td>0.323324</td>\n",
       "      <td>0.346154</td>\n",
       "      <td>0.001017</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46788073</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45516128</td>\n",
       "      <td>on</td>\n",
       "      <td>-1271945.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2925608</th>\n",
       "      <td>rs2074404</td>\n",
       "      <td>lnc_LINC02210_CRHR1_7</td>\n",
       "      <td>6.281516e-08</td>\n",
       "      <td>-0.768667</td>\n",
       "      <td>0.130140</td>\n",
       "      <td>0.281600</td>\n",
       "      <td>0.346154</td>\n",
       "      <td>0.010639</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46788073</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45585004</td>\n",
       "      <td>on</td>\n",
       "      <td>-1203069.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2933454</th>\n",
       "      <td>rs2074404</td>\n",
       "      <td>ENSG00000238083.7</td>\n",
       "      <td>4.553181e-09</td>\n",
       "      <td>0.815159</td>\n",
       "      <td>0.125441</td>\n",
       "      <td>0.321793</td>\n",
       "      <td>0.346154</td>\n",
       "      <td>0.001047</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46788073</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46511511</td>\n",
       "      <td>on</td>\n",
       "      <td>-276562.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2933484</th>\n",
       "      <td>rs2074404</td>\n",
       "      <td>ENSG00000214425.7</td>\n",
       "      <td>3.239323e-10</td>\n",
       "      <td>-0.861242</td>\n",
       "      <td>0.121685</td>\n",
       "      <td>0.360140</td>\n",
       "      <td>0.346154</td>\n",
       "      <td>0.000130</td>\n",
       "      <td>chr17</td>\n",
       "      <td>46788073</td>\n",
       "      <td>chr17</td>\n",
       "      <td>45550335</td>\n",
       "      <td>on</td>\n",
       "      <td>-1237738.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         variant_id           phenotype_id          pval         b      b_se  \\\n",
       "138444     rs823118             lnc_RHEX_8  4.879842e-08 -0.798306  0.133847   \n",
       "417630   rs73038319    lnc_C1QTNF3_AMACR_3  4.492441e-12 -3.231971  0.404354   \n",
       "1103755     rs26431     ENSG00000175749.11  1.059660e-07  0.840778  0.145314   \n",
       "1415204    rs199351               KLHL7_DT  6.943582e-09  0.816349  0.127466   \n",
       "2786097    rs142167          lnc_PLEKHM1_3  1.082803e-10 -1.052603  0.143912   \n",
       "2796934    rs142167          lnc_PLEKHM1_4  3.646967e-09 -0.957676  0.146265   \n",
       "2801832    rs142167  lnc_LINC02210_CRHR1_7  1.056680e-09 -1.011352  0.148288   \n",
       "2807509    rs142167             KANSL1_AS1  1.697743e-07  0.887911  0.156420   \n",
       "2809678    rs142167      ENSG00000238083.7  9.447630e-12  1.094144  0.139643   \n",
       "2809708    rs142167      ENSG00000214425.7  8.185168e-11 -1.053283  0.142831   \n",
       "2878929    rs199502          lnc_PLEKHM1_3  2.304677e-09 -0.979609  0.147333   \n",
       "2889766    rs199502          lnc_PLEKHM1_4  9.029755e-09 -0.927934  0.146229   \n",
       "2894664    rs199502  lnc_LINC02210_CRHR1_7  2.025256e-10 -1.034765  0.144133   \n",
       "2900341    rs199502             KANSL1_AS1  3.307256e-08  0.920575  0.152085   \n",
       "2902510    rs199502      ENSG00000238083.7  8.092927e-11  1.044152  0.141546   \n",
       "2902540    rs199502      ENSG00000214425.7  4.550004e-11 -1.053522  0.140465   \n",
       "2909873   rs2074404          lnc_PLEKHM1_3  4.584136e-10 -0.859158  0.122697   \n",
       "2920710   rs2074404          lnc_PLEKHM1_4  4.108276e-09 -0.801411  0.122894   \n",
       "2925608   rs2074404  lnc_LINC02210_CRHR1_7  6.281516e-08 -0.768667  0.130140   \n",
       "2933454   rs2074404      ENSG00000238083.7  4.553181e-09  0.815159  0.125441   \n",
       "2933484   rs2074404      ENSG00000214425.7  3.239323e-10 -0.861242  0.121685   \n",
       "\n",
       "               r2       maf    bh_fdr  chrom        pos    chr        tss  \\\n",
       "138444   0.285561  0.362637  0.008725   chr1  205754444   chr1  205862522   \n",
       "417630   0.417870  0.021978  0.000014   chr3   18320267   chr5   34190204   \n",
       "1103755  0.273335  0.241758  0.017051   chr5  103030090   chr5  103032376   \n",
       "1415204  0.315474  0.445055  0.001490   chr7   23260430   chr7   23105703   \n",
       "2786097  0.375429  0.296703  0.000058  chr17   46717868  chr17   45520464   \n",
       "2796934  0.325094  0.296703  0.000978  chr17   46717868  chr17   45516128   \n",
       "2801832  0.343246  0.296703  0.000340  chr17   46717868  chr17   45585004   \n",
       "2807509  0.265811  0.296703  0.026017  chr17   46717868  chr17   46193573   \n",
       "2809678  0.408212  0.296703  0.000015  chr17   46717868  chr17   46511511   \n",
       "2809708  0.379276  0.296703  0.000053  chr17   46717868  chr17   45550335   \n",
       "2878929  0.331874  0.269231  0.000674  chr17   46784981  chr17   45520464   \n",
       "2889766  0.311513  0.269231  0.001816  chr17   46784981  chr17   45516128   \n",
       "2894664  0.366736  0.269231  0.000093  chr17   46784981  chr17   45585004   \n",
       "2900341  0.291623  0.269231  0.006261  chr17   46784981  chr17   46193573   \n",
       "2902510  0.379432  0.269231  0.000053  chr17   46784981  chr17   46511511   \n",
       "2902540  0.387277  0.269231  0.000049  chr17   46784981  chr17   45550335   \n",
       "2909873  0.355222  0.346154  0.000164  chr17   46788073  chr17   45520464   \n",
       "2920710  0.323324  0.346154  0.001017  chr17   46788073  chr17   45516128   \n",
       "2925608  0.281600  0.346154  0.010639  chr17   46788073  chr17   45585004   \n",
       "2933454  0.321793  0.346154  0.001047  chr17   46788073  chr17   46511511   \n",
       "2933484  0.360140  0.346154  0.000130  chr17   46788073  chr17   45550335   \n",
       "\n",
       "         onchrom       dist  \n",
       "138444        on   108078.0  \n",
       "417630   between        NaN  \n",
       "1103755       on     2286.0  \n",
       "1415204       on  -154727.0  \n",
       "2786097       on -1197404.0  \n",
       "2796934       on -1201740.0  \n",
       "2801832       on -1132864.0  \n",
       "2807509       on  -524295.0  \n",
       "2809678       on  -206357.0  \n",
       "2809708       on -1167533.0  \n",
       "2878929       on -1264517.0  \n",
       "2889766       on -1268853.0  \n",
       "2894664       on -1199977.0  \n",
       "2900341       on  -591408.0  \n",
       "2902510       on  -273470.0  \n",
       "2902540       on -1234646.0  \n",
       "2909873       on -1267609.0  \n",
       "2920710       on -1271945.0  \n",
       "2925608       on -1203069.0  \n",
       "2933454       on  -276562.0  \n",
       "2933484       on -1237738.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(sig_risk_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    2.000000e+01\n",
       "mean     8.392539e+05\n",
       "std      4.957181e+05\n",
       "min      2.286000e+03\n",
       "25%      2.757890e+05\n",
       "50%      1.182468e+06\n",
       "75%      1.235419e+06\n",
       "max      1.271945e+06\n",
       "Name: dist, dtype: float64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_risk_df['dist'].abs().describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### save the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "trans_df.to_parquet(risk_results_file, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run the cell-type fraction interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# subset just the cell fraction covariates\n",
    "keep_cols = cell_types_list + ['sampleid']\n",
    "cf_covs_df = covs_df[keep_cols]\n",
    "cf_covs_df.set_index('sampleid', inplace=True)\n",
    "print(cf_covs_df.shape)\n",
    "# display(cf_covs_df.head())\n",
    "\n",
    "for cell_type in cell_types_list:\n",
    "    cf_covs_df[cell_type].fillna(cf_covs_df[cell_type].mean(), inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DopaminergicNeurons\n",
      "trans-QTL mapping\n",
      "  * 91 samples\n",
      "  * 30944 phenotypes\n",
      "  * 0 covariates\n",
      "  * 104 variants\n",
      "  * including interaction term\n",
      "    processing batch 1/1\n",
      "    time elapsed: 0.01 min\n",
      "(3187232, 14)\n",
      "EarlyneuronProgenitor\n",
      "trans-QTL mapping\n",
      "  * 91 samples\n",
      "  * 30944 phenotypes\n",
      "  * 0 covariates\n",
      "  * 104 variants\n",
      "  * including interaction term\n",
      "    processing batch 1/1\n",
      "    time elapsed: 0.00 min\n",
      "(6374464, 14)\n",
      "ImmatureDopaminergicNeurons\n",
      "trans-QTL mapping\n",
      "  * 91 samples\n",
      "  * 30944 phenotypes\n",
      "  * 0 covariates\n",
      "  * 104 variants\n",
      "  * including interaction term\n",
      "    processing batch 1/1\n",
      "    time elapsed: 0.00 min\n",
      "(9561696, 14)\n",
      "ProliferatingFloorPlateProgenitors\n",
      "trans-QTL mapping\n",
      "  * 91 samples\n",
      "  * 30944 phenotypes\n",
      "  * 0 covariates\n",
      "  * 104 variants\n",
      "  * including interaction term\n",
      "    processing batch 1/1\n",
      "    time elapsed: 0.00 min\n",
      "(12779872, 14)\n",
      "LateneuronProgenitor\n",
      "trans-QTL mapping\n",
      "  * 91 samples\n",
      "  * 30944 phenotypes\n",
      "  * 0 covariates\n",
      "  * 104 variants\n",
      "  * including interaction term\n",
      "    processing batch 1/1\n",
      "    time elapsed: 0.00 min\n",
      "(15967104, 14)\n",
      "Ependymal-likeCells\n",
      "trans-QTL mapping\n",
      "  * 91 samples\n",
      "  * 30944 phenotypes\n",
      "  * 0 covariates\n",
      "  * 104 variants\n",
      "  * including interaction term\n",
      "    processing batch 1/1\n",
      "    time elapsed: 0.00 min\n",
      "(19154336, 14)\n",
      "Neuroepithelial-likeCells\n",
      "trans-QTL mapping\n",
      "  * 91 samples\n",
      "  * 30944 phenotypes\n",
      "  * 0 covariates\n",
      "  * 104 variants\n",
      "  * including interaction term\n",
      "    processing batch 1/1\n",
      "    time elapsed: 0.00 min\n",
      "(22341568, 14)\n",
      "CPU times: user 3min 52s, sys: 13.2 s, total: 4min 5s\n",
      "Wall time: 3min 55s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# run mapping with interactions for cell types\n",
    "cells_df = None\n",
    "for cell_type in cell_types_list:\n",
    "    print(cell_type)\n",
    "    trans_df = trans.map_trans(genotype_df, phenotype_df, \n",
    "                               covariates_df=pd.DataFrame(index=phenotype_df.columns), \n",
    "                               batch_size=10000, return_r2=True,\n",
    "                               return_sparse=True, pval_threshold=1, maf_threshold=0.01, \n",
    "                               interaction_s=cf_covs_df[cell_type])\n",
    "    \n",
    "    # add the corrected p-value, note just based on all chrom gene pvalues    \n",
    "    trans_df['bh_fdr'] = compute_fdr(trans_df['pval_gi'].fillna(1))\n",
    "    # add physical positions onto results\n",
    "    trans_df = trans_df.merge(variant_df, how='left', \n",
    "                              left_on='variant_id', right_on='snp')\n",
    "    trans_df = trans_df.merge(phenotype_pos_df, how='left', \n",
    "                              left_on='phenotype_id', right_on='phenotype_id')\n",
    "    trans_df['onchrom'] = np.where(trans_df['chrom'] == trans_df['chr'], \n",
    "                                   'on', 'between')\n",
    "    trans_df['dist'] = trans_df['tss'] - trans_df['pos']\n",
    "    trans_df.loc[trans_df['onchrom'] == 'between', 'dist'] = np.nan\n",
    "    trans_df['celltype'] = cell_type\n",
    "    cells_df = pd.concat([cells_df, trans_df])\n",
    "    print(cells_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the results\n",
    "cells_df.to_parquet(risk_cs_results_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6255,)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cells_df.loc[cells_df['pval_gi'] <= min_nominal_alpha]['phenotype_id'].unique().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(38823, 14)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>phenotype_id</th>\n",
       "      <th>pval_g</th>\n",
       "      <th>pval_i</th>\n",
       "      <th>pval_gi</th>\n",
       "      <th>maf</th>\n",
       "      <th>bh_fdr</th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>chr</th>\n",
       "      <th>tss</th>\n",
       "      <th>onchrom</th>\n",
       "      <th>dist</th>\n",
       "      <th>celltype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1007</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>lnc_ZNF460_2</td>\n",
       "      <td>0.011791</td>\n",
       "      <td>0.318458</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.015098</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843</td>\n",
       "      <td>chr19</td>\n",
       "      <td>57304305</td>\n",
       "      <td>between</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DopaminergicNeurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2606</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>lnc_SLC25A51_3</td>\n",
       "      <td>0.005231</td>\n",
       "      <td>0.037849</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.035248</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843</td>\n",
       "      <td>chr9</td>\n",
       "      <td>37872760</td>\n",
       "      <td>between</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DopaminergicNeurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2699</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000213642.3</td>\n",
       "      <td>0.001212</td>\n",
       "      <td>0.352339</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.018825</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843</td>\n",
       "      <td>chr7</td>\n",
       "      <td>64570437</td>\n",
       "      <td>between</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DopaminergicNeurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3411</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000207185.1</td>\n",
       "      <td>0.000825</td>\n",
       "      <td>0.415416</td>\n",
       "      <td>0.000107</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.043576</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843</td>\n",
       "      <td>chr11</td>\n",
       "      <td>118593988</td>\n",
       "      <td>between</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DopaminergicNeurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6328</th>\n",
       "      <td>rs76763715</td>\n",
       "      <td>ENSG00000258643.5</td>\n",
       "      <td>0.004346</td>\n",
       "      <td>0.091108</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.10989</td>\n",
       "      <td>0.023881</td>\n",
       "      <td>chr1</td>\n",
       "      <td>155235843</td>\n",
       "      <td>chr14</td>\n",
       "      <td>23306835</td>\n",
       "      <td>between</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DopaminergicNeurons</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      variant_id       phenotype_id    pval_g    pval_i   pval_gi      maf  \\\n",
       "1007  rs76763715       lnc_ZNF460_2  0.011791  0.318458  0.000014  0.10989   \n",
       "2606  rs76763715     lnc_SLC25A51_3  0.005231  0.037849  0.000070  0.10989   \n",
       "2699  rs76763715  ENSG00000213642.3  0.001212  0.352339  0.000020  0.10989   \n",
       "3411  rs76763715  ENSG00000207185.1  0.000825  0.415416  0.000107  0.10989   \n",
       "6328  rs76763715  ENSG00000258643.5  0.004346  0.091108  0.000033  0.10989   \n",
       "\n",
       "        bh_fdr chrom        pos    chr        tss  onchrom  dist  \\\n",
       "1007  0.015098  chr1  155235843  chr19   57304305  between   NaN   \n",
       "2606  0.035248  chr1  155235843   chr9   37872760  between   NaN   \n",
       "2699  0.018825  chr1  155235843   chr7   64570437  between   NaN   \n",
       "3411  0.043576  chr1  155235843  chr11  118593988  between   NaN   \n",
       "6328  0.023881  chr1  155235843  chr14   23306835  between   NaN   \n",
       "\n",
       "                 celltype  \n",
       "1007  DopaminergicNeurons  \n",
       "2606  DopaminergicNeurons  \n",
       "2699  DopaminergicNeurons  \n",
       "3411  DopaminergicNeurons  \n",
       "6328  DopaminergicNeurons  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sig_risk_df = cells_df.loc[cells_df['bh_fdr'] <= alpha_value]\n",
    "# sig_risk_df = cells_df.loc[cells_df['pval_gi'] <= min_nominal_alpha]\n",
    "print(sig_risk_df.shape)\n",
    "display(sig_risk_df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    3.882300e+04\n",
       "mean     6.916414e-05\n",
       "std      7.752032e-05\n",
       "min      4.462828e-18\n",
       "25%      9.183776e-06\n",
       "50%      3.778521e-05\n",
       "75%      1.031995e-04\n",
       "max      3.019048e-04\n",
       "Name: pval_gi, dtype: float64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_risk_df['pval_gi'].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    2.133000e+03\n",
       "mean     5.195420e+07\n",
       "std      4.392311e+07\n",
       "min      2.070100e+04\n",
       "25%      1.569256e+07\n",
       "50%      3.866040e+07\n",
       "75%      7.946684e+07\n",
       "max      1.902235e+08\n",
       "Name: dist, dtype: float64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_risk_df['dist'].abs().describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ProliferatingFloorPlateProgenitors    3218176\n",
       "EarlyneuronProgenitor                 3187232\n",
       "Neuroepithelial-likeCells             3187232\n",
       "DopaminergicNeurons                   3187232\n",
       "Ependymal-likeCells                   3187232\n",
       "LateneuronProgenitor                  3187232\n",
       "ImmatureDopaminergicNeurons           3187232\n",
       "Name: celltype, dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cells_df['celltype'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ProliferatingFloorPlateProgenitors    19432\n",
       "DopaminergicNeurons                    8965\n",
       "Neuroepithelial-likeCells              5909\n",
       "EarlyneuronProgenitor                  2418\n",
       "LateneuronProgenitor                   1622\n",
       "Ependymal-likeCells                     253\n",
       "ImmatureDopaminergicNeurons             224\n",
       "Name: celltype, dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_risk_df['celltype'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### save the significant results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "sig_risk_df.to_csv(sig_risk_cs_results_file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "name": "pytorch-gpu.1-7.mnightly-2021-02-12-debian-10-test",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/pytorch-gpu.1-7:mnightly-2021-02-12-debian-10-test"
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false,
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
