# Notebooks for generation of draft figures and visualizaitons

## Compare FOUNDIN-PD da65 SCRN-DA to HipSci day 52 SCRN-DA eQTL and FOUNID RNAB to HipSci SCRN pseudo bulk; figures/compare_foundin_hipsci.ipynb

## Visualization of PD risk and QTL colocalization results as a heatmap; figures/colocalization_results_heatmap.ipynb

## Plot feature specific QTL and PD risk results; figures/feature_specific_qtl_results.ipynb and figures/pm_run_feature_specific_qtl_results.ipynb

## Generate locus tracks for modalities and results for loci of interest; figures/locus_tracks.ipynb and figures/pm_run_locus_tracks.ipynb

## Visualize cell-type importance based on colocalziation between PD risk and cell-specific eQTL; figures/cell_enrichment_risk_qtl_colocaliation.ipynb
  
## Generate the GSEA Enrichr results for the specified colocalizations results and enrichment libraries; figures/gsea_enrichr.ipynb




figures/cell_pathway_sankey.ipynb is dependent on gsea stuff

## visual QTS for specific feature; figures/plot_qts_pair.ipynb

## Compare cell-type fractions from single-cell modalities and the SCADEN estimated cell-type fractions for the RNAB; figures/compare_cell_fractions.ipynb

## Check the CpG sites indentifed in EWAS of PD/LBD Lewy Body Pathology in FOUNDIN-PD Epigentic Analyses; figures/lewy_body_path_ewas.ipynb
- Pihlstr√∏m L, Shireby G, Geut H et al. Epigenome-wide association study of human frontal cortex identifies differential methylation in Lewy body pathology. Nat Commun 2022;13:4932. [PUBMED ID: 35995800](https://pubmed.ncbi.nlm.nih.gov/35995800/)

## For monogenic genes to visualize omics modalites at the locus, similar to the locus tracks above for common risk; figures/monogenic_tracks.ipynb and figures/pm_run_monogenic_tracks.ipynb. This was originally migrated from the epigenetics project.
